---
title: "Introduction Page"
format: html
---

scExploreR allows for in-depth exploration of single-cell datasets without code. Functionality for the app is grouped into three tabs, the info tab, the plots tab, and the differential expression tab.

## Info Tab

The information tab is grouped into the following pages, which may be switched using the sidebar on the left:

- **General Information:** Information about the app (current page)
- **Dataset Information:** Information on the currently loaded dataset 
- **Dataset Contents:** Information on metadata variables in the current dataset

## Plots Tab

Use the plots tab to produce and export several plot types commonly used in single-cell analysis. The following visualizations are available:

- DimPlots
- Feature Plots
- Feature Co-expression plots
- Violin Plots
- Dot Plots
- Scatterplots
- Ridge Plots
- Cell Type Proportion Plots

For more information on plot types, go to the "?" button and select "Single-Cell Visualizations". For information on all features in the plots tab, please see the [full documentation](https://amc-heme.github.io/scExploreR/articles/full_documentation.html#plots-tab).

## Differential Expression Tab

Use the differential expression tab to identify genes that are upregulated or downregulated between groups of cells in the dataset. Currently, a Wilcoxon rank sum method is used, with plans to add pseudobulk testing in the near future.

### Running a Differential Expression Test

Use the menu on the left hand side of the app to run a differential gene expression test.

::: {style="width: 600px;"}
!["Diagram on the process of running a DGE test, explained below."](intro_page_dge_figure.svg)
:::

**Step 1: Choose test**

- **Differential expression:** Compare two cell groups to find genes enriched in one group versus the other.  
- **Marker identification:** Split cells into three or more groups to find genes enriched in each group versus all others.  

**Step 2: Define groups**  

- **For Marker Identification:** choose a metadata variable. Groups are automatically created for all values of that variable.
- **For Differential expression:** can define using metadata or feature expression.
  - **By Metadata**: select a metadata variable, and then choose two values from that variable from which to form groups.
  - **By feature expression**:
    1. Select a feature.  
    2. A ridge plot interface will appear. Select a expression threshold by clicking within the plot.
    3. Cells â‰¥ threshold are assigned to Group 1, cells < threshold are assigned to Group 2.  
    
**Step 3: Select analysis scope**

- **Full object:** Run on all cells.  
- **Subset:** Apply filters based on metadata or feature expression, then run on the filtered cells.

For more information, see the [full documentation](https://amc-heme.github.io/scExploreR/articles/full_documentation.html#dge-tab).

## Subsetting

Subsetting can be performed in both the plots tab and the differential expression tab using the "Subset Options" interface. Use this interface to define a set of criteria for subsetting. Criteria can be defined around metadata or feature expresson, and the cells that match the criteria are included in the subset. 

To define criteria, open the "Subset Options" dropdown and follow the instructions below. Any number of filter criteria can be defined, and the criteria are joined via AND logic (cells matching criterion A, and criterion B, and so on are included). When finished, use the **"Apply Subset"** button to create the subset.

### Adding filter crieria

1. Select "Add filter"
2. Choose the type of filter, either **"Categorical Metadata"**, **"Feature Expression"** (which includes numeric metadata), or **"Advanced Subsetting"**.
3. Define the filter criterion based on the type selected in 2):
  - **"Categorical Metadata":** Select a metadata variable, and select the values to include for that variable.
  - **"Feature Expression":** Select a feature or numeric metadata variable. A ridge plot interface will appear. Use the ">", "<", and "Range" buttons to filter for cells with values greater than, less than, or within a range of, the selected value(s), respectively. 
    - For less than or greater than filters, select a single value on the plot that will be used as the threshold. 
    - For the range filter, select an upper and lower bound on the plot, and all cells with values between the values selected will be included. To edit either bound after selecting them on the plot, select the pencil icon beneath the plot.
  - **"Advanced Subsetting":** This is only available for Seurat objects. Advanced users may define subsets by code by passing a string that is interperatable to the `subset` argument for the `subset` function in Seurat.
4. Press "Confirm Filter" to apply the filter (or the "Cancel" button to cancel). 

### Editing or deleting filters

Each filter criterion displayed has a pencil (<i class="fa fa-pencil"></i>) and an "X" (<i class="fa fa-times-circle"></i>) button on the lower right hand side. Click the pencil button to edit a filter, and the "X" button to delete it. 

## Datasets

To view the datasets available in this deployment, go to the <span class="options-button-representation"><i class="fas fa-ellipsis"></i></span> button, and select "Choose Dataset". Select a dataset from the widow that appears to change datasets.

