<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scExploreR Config App Documentation • scExploreR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scExploreR Config App Documentation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scExploreR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Help and Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/dataset_setup_walkthrough.html">App Setup Walkthrough</a></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">Exploring your Data</a></li>
    <li><a class="dropdown-item" href="../articles/object_guidelines.html">Object Preparation Guidelines</a></li>
    <li><a class="dropdown-item" href="../articles/docker.html">Deploying With Docker</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Additional Information</h6></li>
    <li><a class="dropdown-item" href="../articles/scRNA_Plots_Explained.html">Guide to scRNA-Seq Plots</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_subsetting_documentation.html">String Subsetting Guide</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Full Documentation</h6></li>
    <li><a class="dropdown-item" href="../articles/full_documentation.html">scExploreR Documentation</a></li>
    <li><a class="dropdown-item" href="../articles/config_documentation.html">Config App Documentation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/amc-heme/scExploreR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="config_documentation_files/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="config_documentation_files/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scExploreR Config App Documentation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amc-heme/scExploreR/blob/main/vignettes/config_documentation.Rmd" class="external-link"><code>vignettes/config_documentation.Rmd</code></a></small>
      <div class="d-none name"><code>config_documentation.Rmd</code></div>
    </div>

    
    
<style type="text/css">
.center{text-align:center;display:block}.white-text{color:white}.card-header{background-color:#CFB87C;margin-top:0px}.card-body{background-color:#E1E1E1;border-radius:0px 0px var(--bs-card-border-radius) var(--bs-card-border-radius)}.callout{padding:20px;margin:20px 0;border:1px solid #eee;border-left-width:8px;border-radius:8px}.callout h4,.callout h5{margin-top:0;margin-bottom:5px}.callout p:last-child{margin-bottom:0}.callout code{border-radius:3px}.callout+.bs-callout{margin-top:-5px}.callout-default{border-left-color:#777;background-color:#77735}.callout-default h4,.callout-default h5{color:#777}.callout-primary{border-left-color:#428bca;background-color:#428bca35}.callout-primary h4,.callout-primary h5{color:#428bca}.callout-success{border-left-color:#5cb85c;background-color:#5cb85c35}.callout-success h4,.callout-success h5{color:#5cb85c}.callout-danger{border-left-color:#d9534f;background-color:#d9534f35}.callout-danger h4,.callout-danger h5{color:#d9534f}.callout-warning{border-left-color:#f0ad4e;background-color:#f0ad4e35}.callout-warning h4,.callout-warning h5{color:#f0ad4e}.callout-info{border-left-color:#5bc0de;background-color:#5bc0de35}.callout-info h4,.callout-info h5{color:#5bc0de}.callout-bdc{border-left-color:#29527a;background-color:#29527a35}.callout-bdc h4,.callout-bdc h5{color:#29527a}.scroll-80{max-height:80px;overflow-y:auto;background-color:inherit}.scroll-100{max-height:100px;overflow-y:auto;background-color:inherit}.scroll-300{max-height:300px;overflow-y:auto;background-color:inherit}
</style>
<style type="text/css"> 
  .centered-image{
    display: block; 
    margin-left: auto; 
    margin-right: auto;
    } 
    </style>
<p>The scExploreR config app is used to configure single-cell objects
for display in the main browser. This vignette will walk through the
operation of the app and detail all settings that may be adjusted.</p>
<p>For guidelines on preparing objects before loading them into the
config app, see the <a href="object_guidelines.html">object guidelines
vignette</a>, and see the <a href="dataset_setup_walkthrough.html">app
setup walkthrough vignette</a> for an interactive walkthrough of both
object preparation and use of the config app.</p>
<div class="section level2">
<h2 id="running-the-config-app">Running the config app<a class="anchor" aria-label="anchor" href="#running-the-config-app"></a>
</h2>
<p>The config app can be ran from R Studio or the console by entering
the following command:</p>
<pre><code><span><span class="fu"><a href="../reference/run_config.html">run_config</a></span><span class="op">(</span></span>
<span>  object_path <span class="op">=</span> <span class="st">"./object.rds"</span>, </span>
<span>  config_path <span class="op">=</span> <span class="cn">NULL</span> </span>
<span>  <span class="op">)</span></span></code></pre>
<p>The <code>object_path</code> argument is required. It must be an set
to one of the following:</p>
<ul>
<li>For Seurat objects, the path to the .rds file.</li>
<li>For SingleCellExperiment objects, the path to the .rds file. If the
object has been saved via
<code><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html" class="external-link">HDF5Array::saveHDF5SummarizedExperiment()</a></code>, this should be
the path to the <strong>directory</strong> created.</li>
<li>For anndata objects, the path to the .h5ad file.</li>
</ul>
<p>The <code>config_path</code> argument is optional and is used to load
an existing config file into the app for further editing (for more on
this, see <a href="#load-config-file">loading a config file</a>
below).</p>
<div class="callout callout-warning">
<h5>
Warning
</h5>
<p>If you are loading a SingleCellExperiment saved using the
<code>HDF5Array</code> package, you must additionally provide the
parameter <code>is_HDF5SummarizedExperiment = TRUE</code>. The config
app will not load successfully unless you do this.</p>
</div>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>Make sure that the config file loaded is based on the same object as
the one specified. We recommend naming your config files using the
convention “{object_name}_config.yaml” to avoid confusion.</p>
</div>
</div>
<div class="section level2">
<h2 id="general-info-tab">General Info Tab<a class="anchor" aria-label="anchor" href="#general-info-tab"></a>
</h2>
<p>Upon loading the config app, the “General” tab will be displayed.
This tab is used to enter information on the dataset that will be
displayed in the main app when the user opens the “Choose Dataset”
window.</p>
<p>Use <strong>“Label for Dataset”</strong> to enter a short label for
the dataset. Add detailed info for the dataset using the
<strong>“Description of Dataset”</strong> field.</p>
<p><img src="config_documentation-001-Label_and_description.png"></p>
<p>Use the <strong>“Content for Dataset Preview”</strong> menu to
optionally add a plot as a preview image in the “Choose Dataset” window
of the main app. When selecting “DimPlot” as a preview option, a window
will appear below with settings for the plot, and a window showing the
plot as it will appear in the main app.</p>
<div class="section level5">
<h5 id="plot-settings">Plot Settings<a class="anchor" aria-label="anchor" href="#plot-settings"></a>
</h5>
<p>Use the <strong>“Metadata to Group By”</strong> to choose the
variable for coloring the cells. If the “Label Groups” checkbox is
enabled, each color group will also be labeled with the value it
represents.The <strong>“Metadata to Split By”</strong> menu may
optionally be used to select a variable to “split” the plot, and the
number of columns in a split plot can be controlled with the
<strong>“Number of columns”</strong> slider. The projection used for the
plot is set with <strong>“Choose Reduction”</strong>.</p>
<p><img src="config_documentation-002-DimPlot_Preview.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="assays-tab">Assays Tab<a class="anchor" aria-label="anchor" href="#assays-tab"></a>
</h2>
<p>Settings for the assays to be included in the Seurat object are
entered in the assays tab.</p>
<div class="section level5">
<h5 id="addingremoving-assays">Adding/Removing Assays<a class="anchor" aria-label="anchor" href="#addingremoving-assays"></a>
</h5>
<p>To choose which assays to expose in the browser, use the
<strong>“Choose assays to include”</strong> menu in the left half of the
window. All assays added to the object will appear on the left hand side
of the two-column menu, under <strong>“Available Assays”</strong>.
Select an assay from this tab to add it to the browser. The assay will
now appear under the <strong>“Selected Assays”</strong> column, and a
panel of options specific to the assay will appear on the right half of
the app window. To remove an assay from the browser, click the assay
name in the <strong>“Selected Assays”</strong> tab.</p>
<p><img src="config_documentation-003-Add-Remove-assays.gif" width="690"></p>
</div>
<div class="section level4">
<h4 id="general-options">General Options<a class="anchor" aria-label="anchor" href="#general-options"></a>
</h4>
<p>On the upper right-hand corner of the screen, a panel of general
assay-related options is displayed. The available options are explained
below.</p>
<!-- May add a class for displaying headers that correspond to input labels in 
     the app -->
<div class="section level5">
<h5 id="choose-genes-assay">
<em>“</em>Choose Genes Assay”<a class="anchor" aria-label="anchor" href="#choose-genes-assay"></a>
</h5>
<p>The assay selected in this menu will be the assay used when computing
differential expression. <!--or correlation;--> It does not effect
plotting.</p>
<p>Note, if you wish to include only non-gene assays in the object, you
may designate those assays as the “genes” assay, but this may cause
issues with the DGE and correlation analyses. The analyses are validated
for gene expression only, and may not be accurate when using other data
types. We will address means of including these data types in analyses
in the future.</p>
</div>
<div class="section level5">
<h5 id="choose-adt-assay">“Choose ADT Assay”<a class="anchor" aria-label="anchor" href="#choose-adt-assay"></a>
</h5>
<p>If you have a surface protein (ADT) assay in your object, you can
specify the assay here to define ADT thresholds. ADT thresholds
transform expression data for plots to filter out cells that are
unlikely to express the ADT in reality, improving the clarity of feature
plots. For more information, see the “ADT Threshold Tab” section.</p>
</div>
</div>
<div class="section level4">
<h4 id="options-for-each-assay">Options for Each Assay<a class="anchor" aria-label="anchor" href="#options-for-each-assay"></a>
</h4>
<p>Each assay-specific panel will have a text box for entering a label,
and a checkbox for including the name of the assay on plots. The label
appears in the feature selection menus, and it may be set to a
descriptive name to improve the presentation of the main browser
(i.e. displaying the genes assay as “Genes”, or the ADT assay as
“Surface Protein”, rather than using the Seurat defaults “RNA” and
“ADT”). If the checkbox is selected, the label entered will display on
plots. For example, if the label is “Gene”, feature plots for the gene
“TP53” will have a default title of “TP53 (Gene)” instead of just
“TP53”.</p>
<p><img src="config_documentation-004-Assay-options.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="metadata-tab">Metadata Tab<a class="anchor" aria-label="anchor" href="#metadata-tab"></a>
</h2>
<p>The metadata tab is used to specify which metadata variables to
include, and to set settings for each metadata variable.</p>
<div class="section level5">
<h5 id="addingremovingsorting-metadata">Adding/Removing/Sorting Metadata<a class="anchor" aria-label="anchor" href="#addingremovingsorting-metadata"></a>
</h5>
<p>Use the <strong>“Choose Metadata to Include”</strong> menu on the
left side of the screen to choose metadata variables to include in the
browser. Initially, all variables in <code>object@meta.data</code> will
display under the <strong>“Available Metadata”</strong> column. To
include a metadata variable, click and drag the variable to the
<strong>“Included Metadata”</strong> section.</p>
<p><strong>Which metadata should I include?</strong> We recommend
including variables that are directly related to the experimental design
and are easily understood by your target audience. Variables describing
properties such as cell type, sample, or patient, or those related to
experimental design such as control vs. treated cells, or timepoint,
should be included in the browser. It is not recommended to include
variables such as the capture chemistry used in scRNA-seq, or cell type
annotations that are not biologically relevant (such as numbered
clusters applied by Seurat, unless you have not identified these
clusters or you have determined they are relevant to your target
audience).</p>
<p>The order the variables appear in this column is recognized by the
browser; metadata in the plotting and subset menus will appear in the
order set here. <strong>The first variable will be the default variable
used for coloring plots</strong>; we highly recommend setting this to
“clusters”, “cell type”, or a similar variable. To remove a variable,
drag it back to the <strong>“Available Metadata”</strong> column.</p>
<p><img src="config_documentation-005-Metadata-Include-Exclude.gif" width="690"></p>
</div>
<div class="section level4">
<h4 id="general-options-1">General Options<a class="anchor" aria-label="anchor" href="#general-options-1"></a>
</h4>
<div class="section level5">
<h5 id="patientsample-variable">“Patient/Sample Variable”<a class="anchor" aria-label="anchor" href="#patientsample-variable"></a>
</h5>
<p>The variable chosen in this selection menu will be the one used for
the “Metadata Pie Chart”
<!-- Link to metadata pie chart documentation --> plot in the main app.
If your dataset consists of multiple samples (or patients), choose the
metadata variable that corresponds to sample or patient to allow users
to view composition of the dataset at the sample level (for example,
seeing how many patients are in one treatment category or time point
versus another, as opposed to how many cells).</p>
</div>
</div>
<div class="section level4">
<h4 id="options-for-each-metadata-variable">Options for Each Metadata Variable<a class="anchor" aria-label="anchor" href="#options-for-each-metadata-variable"></a>
</h4>
<p>A panel of options will appear on the right side of the screen for
each metadata variable included. The header of each panel will display
the name of the variable as it appears in the object, whether the
variable is categorical or numeric, and how many unique values of the
variable exist. Beneath the header, unique values are displayed to
assist in the identification and naming of variables.</p>
<p>Currently, only categorical metadata is available for inclusion in
the object, since numeric metadata does not yet work with the subset
menus. Support for numeric metadata will be added in a future
release.</p>
<p>Use the <strong>“Set label for metadata variable”</strong> text box
to set a display name in the app. The name entered here will display in
all plotting menus that use categorical metadata, as well as in subset
menus in all three tabs.</p>
<p>The <strong>“Group metadata into categories?”</strong> switch is used
to create display groups using the values in the metadata variable. The
groups chosen here are displayed in the main app as a guide for choosing
subsets in either of the tabs of the main app. This is useful in cases
where the dataset contains samples from different experimental groups.
Grouping samples in the config app aids the end user in understanding
which samples fall in which treatment group. Groups may also be used for
clustering detailed cell annotations into more general annotations for
facilitated interpretation by the end user.</p>
<p><img src="config_documentation-006-Metadata-Groups-Interface.png" width="371"></p>
<div class="section level5">
<h5 id="using-the-metadata-groups-interface">Using the Metadata Groups Interface<a class="anchor" aria-label="anchor" href="#using-the-metadata-groups-interface"></a>
</h5>
<p>Enable the <strong>“Group metadata into categories?”</strong> switch
to display the interface. Two fields will display initially for the
first group.</p>
<p>The first field is used to choose a name for the group, and the
second field is used to choose group members from the classes of the
variable. Available classes will appear as you type.</p>
<p>To add additional groups, press the “Add Group” button. For each
group added, fields for the group name and members will appear.</p>
<p>To delete a group, click the “X” button to the right of each field.
The first field can’t be deleted; to omit this field, disable the
<strong>“Group metadata into categories?”</strong> switch.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="reductions-tab">Reductions Tab<a class="anchor" aria-label="anchor" href="#reductions-tab"></a>
</h2>
<p>The reductions tab is used to set the dimensional reductions
(projections) to include in the object, and in which order to display
them. The selections here affect all DimPlots and Feature Plots in the
app.</p>
<p>Click and drag reductions from the <strong>“Available
Reductions”</strong> to the <strong>“Included Reductions”</strong>
section to include them in the app. The order the reductions appear will
be the order they will display in the in-app menus, and the first
reduction in the config app will be the default projection displayed in
the app. Click and drag the reductions to change their order.</p>
<p>For each reduction, the <strong>label</strong> may be set. The
reduction will appear in the app as it is labeled here.</p>
</div>
<div class="section level2">
<h2 id="adt-threshold-tab">ADT Threshold Tab<a class="anchor" aria-label="anchor" href="#adt-threshold-tab"></a>
</h2>
<p>For Seurat Objects with ADT assays, the ADT Threshold Tab allows the
user to choose thresholds below which the biological expression of ADTs
is likely zero, even if the expression values themselves is greater than
zero. The use of this tab will help remove statistical noise from the
measurements, giving the end user a more accurate picture of where
surface proteins are being expressed in a sample.</p>
<div class="section level4">
<h4 id="enabling-the-adt-threshold-tab">Enabling the ADT Threshold Tab<a class="anchor" aria-label="anchor" href="#enabling-the-adt-threshold-tab"></a>
</h4>
<p>The ADT Threshold Tab is hidden by default. To show the tab, make a
selection under “Choose ADT Assay” in the assays tab.</p>
<p><img src="config_documentation-007-Enable-ADT-Threshold.gif" width="690"></p>
</div>
<div class="section level4">
<h4 id="adding-adt-thresholds">Adding ADT Thresholds<a class="anchor" aria-label="anchor" href="#adding-adt-thresholds"></a>
</h4>
<p>To add an ADT threshold, click the <strong>“New Threshold”</strong>
button in the ADT threshold tab, and then use search menu that appears
to the right of the table to search for an ADT.</p>
<p>When an ADT is selected in the search menu, a ridge plot showing the
distribution of expression values in each cell will appear below the
menu. Click the plot to set a threshold for the ADT. To aid in placement
of the threshold, a grey vertical line will appear when mousing over the
plot, and a darker line will appear at the last value selected.</p>
<p>When a selection is made, a panel of statistics will display below
the plot. The panel will show the value of the chosen threshold, and the
number and percentage of cells above and below the threshold.</p>
<p>To save a threshold, click the <strong>“Confirm”</strong> button
beneath the stats panel. To exit without saving the threshold, click
<strong>“Cancel”</strong>. When a threshold is saved, it will appear on
the table of saved thresholds on the left hand side of the screen.</p>
<p><img src="config_documentation-008-Add-ADT-Threshold.gif" width="690"></p>
</div>
<div class="section level4">
<h4 id="editing-and-deleting-adt-thresholds">Editing and Deleting ADT Thresholds<a class="anchor" aria-label="anchor" href="#editing-and-deleting-adt-thresholds"></a>
</h4>
<p>To edit an existing threshold, click the
<i class="fas fa-pencil" role="presentation" aria-label="pencil icon"></i>
icon next to the desired ADT in the table. The same interface shown when
adding an ADT value will appear on the right hand side of the screen,
with the previous value selected. Press <strong>“Confirm”</strong> to
save any changes made.</p>
<p>To delete a threshold, press the
<i class="far fa-circle-xmark" role="presentation" aria-label="circle-xmark icon"></i>
button next to the desired ADT.</p>
<p><img src="config_documentation-009-Edit-ADT-Threshold.gif" width="690"></p>
<p><img src="config_documentation-010-Delete-ADT-Threshold.gif" width="690"></p>
</div>
</div>
<div class="section level2">
<h2 id="saving-and-loading-config-files">Saving and Loading Config Files<a class="anchor" aria-label="anchor" href="#saving-and-loading-config-files"></a>
</h2>
<div class="section level4">
<h4 id="saving">Saving<a class="anchor" aria-label="anchor" href="#saving"></a>
</h4>
<p>To save a config file, press the
<i class="fas fa-ellipsis" role="presentation" aria-label="ellipsis icon"></i>
button in the upper right corner and select “Save Config File”. A window
will appear to save the config file to your local file system. The file
will save in a .yaml format.</p>
</div>
<div class="section level4">
<h4 id="load-config-file">Loading<a class="anchor" aria-label="anchor" href="#load-config-file"></a>
</h4>
<p>To load a config file, call <code><a href="../reference/run_config.html">run_config()</a></code> with the path
to the saved file:</p>
<pre><code><span><span class="fu"><a href="../reference/run_config.html">run_config</a></span><span class="op">(</span></span>
<span>  object_path <span class="op">=</span> <span class="st">"./object.rds"</span>, </span>
<span>  config_path <span class="op">=</span> <span class="st">"./config_file.yaml"</span> </span>
<span>  <span class="op">)</span></span></code></pre>
<p>When the app loads, press the
<i class="fas fa-ellipsis" role="presentation" aria-label="ellipsis icon"></i>
button in the upper right corner and select “Load Config File”. All
fields in the app will update based on the information read from the
file. Each time this button is pressed, the information entered will be
reset to what is in the file, so be careful to avoid overwriting data.
If the app is launched with <code>config_path</code> equal to
<code>NULL</code>, the app will not fill any fields.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hematology Informatics Group.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
