<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scExploreR">
<title>Hosting a Dataset in scExploreR and Performing Analysis • scExploreR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Hosting a Dataset in scExploreR and Performing Analysis">
<meta property="og:description" content="scExploreR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scExploreR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Help and Tutorials</h6>
    <a class="dropdown-item" href="../articles/dataset_setup_walkthrough.html">App Setup Walkthrough</a>
    <a class="dropdown-item" href="../articles/tutorial.html">Exploring your Data</a>
    <a class="dropdown-item" href="../articles/scRNA_Plots_Explained.html">Guide to scRNA-Seq Plots</a>
    <a class="dropdown-item" href="../articles/advanced_subsetting_documentation.html">String Subsetting Guide</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Reference</h6>
    <a class="dropdown-item" href="../articles/full_documentation.html">Full App Manual</a>
    <a class="dropdown-item" href="../articles/config_documentation.html">Config App Documentation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/amc-heme/scExploreR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="dataset_setup_walkthrough_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="dataset_setup_walkthrough_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="dataset_setup_walkthrough_files/datatables-binding-0.32/datatables.js"></script><link href="dataset_setup_walkthrough_files/dt-core-bootstrap5-1.13.6/css/dataTables.bootstrap5.min.css" rel="stylesheet">
<script src="dataset_setup_walkthrough_files/dt-core-bootstrap5-1.13.6/js/jquery.dataTables.min.js"></script><script src="dataset_setup_walkthrough_files/dt-core-bootstrap5-1.13.6/js/dataTables.bootstrap5.min.js"></script><script src="dataset_setup_walkthrough_files/bs3compat-0.6.1/transition.js"></script><script src="dataset_setup_walkthrough_files/bs3compat-0.6.1/tabs.js"></script><script src="dataset_setup_walkthrough_files/bs3compat-0.6.1/bs3compat.js"></script><link href="dataset_setup_walkthrough_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="dataset_setup_walkthrough_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Hosting a Dataset in scExploreR and Performing Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amc-heme/scExploreR/blob/HEAD/vignettes/dataset_setup_walkthrough.Rmd" class="external-link"><code>vignettes/dataset_setup_walkthrough.Rmd</code></a></small>
      <div class="d-none name"><code>dataset_setup_walkthrough.Rmd</code></div>
    </div>

    
    
<style type="text/css">
.center{text-align:center;display:block}.white-text{color:white}.card-header{background-color:#CFB87C;margin-top:0px}.card-body{background-color:#E1E1E1;border-radius:0px 0px var(--bs-card-border-radius) var(--bs-card-border-radius)}.callout{padding:20px;margin:20px 0;border:1px solid #eee;border-left-width:8px;border-radius:8px}.callout h4,.callout h5{margin-top:0;margin-bottom:5px}.callout p:last-child{margin-bottom:0}.callout code{border-radius:3px}.callout+.bs-callout{margin-top:-5px}.callout-default{border-left-color:#777;background-color:#77735}.callout-default h4,.callout-default h5{color:#777}.callout-primary{border-left-color:#428bca;background-color:#428bca35}.callout-primary h4,.callout-primary h5{color:#428bca}.callout-success{border-left-color:#5cb85c;background-color:#5cb85c35}.callout-success h4,.callout-success h5{color:#5cb85c}.callout-danger{border-left-color:#d9534f;background-color:#d9534f35}.callout-danger h4,.callout-danger h5{color:#d9534f}.callout-warning{border-left-color:#f0ad4e;background-color:#f0ad4e35}.callout-warning h4,.callout-warning h5{color:#f0ad4e}.callout-info{border-left-color:#5bc0de;background-color:#5bc0de35}.callout-info h4,.callout-info h5{color:#5bc0de}.callout-bdc{border-left-color:#29527a;background-color:#29527a35}.callout-bdc h4,.callout-bdc h5{color:#29527a}.scroll-80{max-height:80px;overflow-y:auto;background-color:inherit}.scroll-100{max-height:100px;overflow-y:auto;background-color:inherit}.scroll-300{max-height:300px;overflow-y:auto;background-color:inherit}
</style>
<!-- Load datasets for use in the vignette --><p>This vignette will walk you through the process of setting up a
dataset in scExploreR. This vignette is based on the Small Cell Lung
Carcinoma dataset analyzed by <a href="https://www.cell.com/cancer-cell/fulltext/S1535-6108(21)00497-9#gr1" class="external-link">Chan
et al. 2021</a>. To download the datasets for this tutorial, <a href="https://cellxgene.cziscience.com/collections/62e8f058-9c37-48bc-9200-e767f318a8ec" class="external-link">go
here</a> and download the following datasets:</p>
<!-- Why is the user downloading these datasets? To replicate the paper. Add a note for this 

Only the combined samples dataset and the SCLC epithelial cells dataset will be covered in the vignette.
-->
<ul>
<li>Combined samples <!-- -   Epithelial cells -->
</li>
<li>SCLC epithelial cells</li>
</ul>
<p>Chan et al. use all three datasets in their analysis, and the same
will be done in this example. The “Epithelial cells” dataset is a subset
of the “Combined samples” dataset that has been re-clustered. Likewise,
the SCLC epithelial cells dataset is a subset of the epithelial cells
dataset. See <strong>Figure 1A</strong> of Chan et al. for a visual
representation.</p>
<p>You may choose either the Anndata or the Seurat formats for download,
though we recommend the Anndata format due to improved performance and
decreased RAM requirements. This article will cover operations on the
Anndata object.</p>
<div class="callout callout-info">
<h4>
Note
</h4>
<p>While the anndata format is exclusive to python, anndata objects can
be explored in R Studio with R-style commands using the <a href="https://rstudio.github.io/reticulate/" class="external-link">Reticulate</a> package. The
example in this vignette uses Reticulate, along with the <a href="https://anndata.dynverse.org" class="external-link">anndata</a> R package, and SCUBA,
which provides utility functions for common exploration/data accession
operations in single cell objects.</p>
</div>
<!-- Vignette Overview/Goals Statement -->
<!-- Update this after finishing the vignette -->
<p>By the end of this example, you will be able to set up an scExploreR
instance with both
datasets<!--, and you will be able to reproduce some of the figures in the paper using the plots tab. This vignette will also show how to identify differentially expressed genes between groups of cells using a wilcoxon rank sum-->.</p>
<div class="card mb-3">
<div class="card-header">
<h4 class="card-title center">
Process to host a dataset in scExploreR
</h4>
</div>
<div class="card-body">
<ol style="list-style-type: decimal">
<li>Explore object to determine appropriate data to display in app.</li>
<li>Use the config app to select relevant data and generate a config
file</li>
<li>Launch scExploreR using the created config file</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="dataset-exploration">Dataset Exploration<a class="anchor" aria-label="anchor" href="#dataset-exploration"></a>
</h2>
<p>To host a dataset in scExploreR, you must indicate which metadata
variables, reductions, and modalities to include in the app. Whether you
are setting up the app for end users or yourself, careful selection of
the most relevant data is key to maximizing the effectiveness of
scExploreR for your dataset.</p>
<p>To understand what aspects to include, it is best to explore the
object to identify relevant modalities, metadata, and reductions. When
hosting objects you have preprocesssed, you will likely be familiar
enough with the object to skip this step.</p>
<p>In this section, we will load the “Combined Samples” and “SCLC
epithelial cells” objects and explore relevant data to include in the
scExploreR instance.</p>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>The example below is one of many ways this step can be approached.
With other objects, the process of understanding relevant data may also
involve communicating with staff involved in sequencing of your samples
and preprocessing of the data.</p>
</div>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>For objects that you preprocess, it is good practice to include
details on the metadata in your object. This is helpful for when you
share your object upon publication, or for sharing your object
internally or with collaborators. There are many ways to do this, though
anndata formats make this easy using the unstructured (<code>uns</code>)
data slot. One suggestion is to add a python dictionary to
<code>uns</code> with relevant variable names as keys, and information
on each variable as values.</p>
</div>
<p>Run the code below to load the Combined Samples dataset, replacing
the path with the path to the downloaded object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://anndata.dynverse.org" class="external-link">anndata</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">full_dataset</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">anndata</span><span class="fu">::</span><span class="fu"><a href="https://anndata.dynverse.org/reference/read_h5ad.html" class="external-link">read_h5ad</a></span><span class="op">(</span></span>
<span>    <span class="co"># Replace the path below with the path to the downloaded dataset</span></span>
<span>    <span class="st">"~/datasets/small_cell_lung_cancer/SCLC_Full.h5ad"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>Metadata variables are chosen when setting up an object to populate
dropdowns with options for grouping (coloring) cells on plots, or for
splitting plots. Chosen metadata also populates the choices avaliable in
subsetting, determining how the user can select sub-populations for
further exploration.</p>
<p>scExploreR allows you to select as many metadata variables as you
like, even all of the ones in the object, but the app functions best
when included metadata is limited to a handful of relevant variables.
Adding variables less relevant to the end user provides limited benefit,
and may even overwhelm users.</p>
<p>The object in this example contains 35 metadata variables. Let’s
explore the variables to narrow down which are the most relevant.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/meta_varnames.html" class="external-link">meta_varnames</a></span><span class="op">(</span><span class="va">full_dataset</span><span class="op">)</span></span></code></pre></div>
<pre class="scroll-100"><code><span><span class="co">##  [1] "ngenes"                                  </span></span>
<span><span class="co">##  [2] "libsize"                                 </span></span>
<span><span class="co">##  [3] "mito_frac"                               </span></span>
<span><span class="co">##  [4] "RBP_frac"                                </span></span>
<span><span class="co">##  [5] "batch"                                   </span></span>
<span><span class="co">##  [6] "donor_id"                                </span></span>
<span><span class="co">##  [7] "treatment"                               </span></span>
<span><span class="co">##  [8] "procedure"                               </span></span>
<span><span class="co">##  [9] "histo"                                   </span></span>
<span><span class="co">## [10] "cell_type_coarse"                        </span></span>
<span><span class="co">## [11] "cell_type_fine"                          </span></span>
<span><span class="co">## [12] "cell_type_general"                       </span></span>
<span><span class="co">## [13] "clusters"                                </span></span>
<span><span class="co">## [14] "cell_type_med"                           </span></span>
<span><span class="co">## [15] "H_knn"                                   </span></span>
<span><span class="co">## [16] "development_stage_ontology_term_id"      </span></span>
<span><span class="co">## [17] "sex_ontology_term_id"                    </span></span>
<span><span class="co">## [18] "self_reported_ethnicity_ontology_term_id"</span></span>
<span><span class="co">## [19] "assay_ontology_term_id"                  </span></span>
<span><span class="co">## [20] "is_primary_data"                         </span></span>
<span><span class="co">## [21] "tissue_ontology_term_id"                 </span></span>
<span><span class="co">## [22] "disease_ontology_term_id"                </span></span>
<span><span class="co">## [23] "organism_ontology_term_id"               </span></span>
<span><span class="co">## [24] "cell_type_ontology_term_id"              </span></span>
<span><span class="co">## [25] "suspension_type"                         </span></span>
<span><span class="co">## [26] "HTAN_Biospecimen_ID"                     </span></span>
<span><span class="co">## [27] "HTAN_Participant_ID"                     </span></span>
<span><span class="co">## [28] "cell_type"                               </span></span>
<span><span class="co">## [29] "assay"                                   </span></span>
<span><span class="co">## [30] "disease"                                 </span></span>
<span><span class="co">## [31] "organism"                                </span></span>
<span><span class="co">## [32] "sex"                                     </span></span>
<span><span class="co">## [33] "tissue"                                  </span></span>
<span><span class="co">## [34] "self_reported_ethnicity"                 </span></span>
<span><span class="co">## [35] "development_stage"</span></span></code></pre>
<p>Only categorical metadata variables may be selected for use in the
app. Below we identify categorical metadata variables.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_unique_vars</span> <span class="op">&lt;-</span> <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/meta_varnames.html" class="external-link">meta_varnames</a></span><span class="op">(</span><span class="va">full_dataset</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Iterate through metadata variables and return a vector specifying whether </span></span>
<span><span class="co"># the variable is categorical (character, factor), or numeric (numeric, </span></span>
<span><span class="co"># double, integer)</span></span>
<span><span class="va">var_classes</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span></span>
<span>    <span class="va">full_unique_vars</span>,</span>
<span>    <span class="co"># Compute the class of the variable</span></span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">meta_var</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sclc_dataset</span><span class="op">$</span><span class="va">obs</span><span class="op">[[</span><span class="va">meta_var</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Sort to group classes together</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">var_classes</span><span class="op">)</span></span></code></pre></div>
<pre class="scroll-300"><code><span><span class="co">##                                    batch </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                 donor_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                treatment </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                procedure </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                    histo </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                 clusters </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##       development_stage_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                     sex_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">## self_reported_ethnicity_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                   assay_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                  tissue_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                 disease_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                organism_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##               cell_type_ontology_term_id </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                          suspension_type </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                      HTAN_Biospecimen_ID </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                      HTAN_Participant_ID </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                cell_type </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                    assay </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                  disease </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                 organism </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                      sex </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                                   tissue </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                  self_reported_ethnicity </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                        development_stage </span></span>
<span><span class="co">##                                 "factor" </span></span>
<span><span class="co">##                          is_primary_data </span></span>
<span><span class="co">##                                "logical" </span></span>
<span><span class="co">##                                   ngenes </span></span>
<span><span class="co">##                                   "NULL" </span></span>
<span><span class="co">##                                  libsize </span></span>
<span><span class="co">##                                   "NULL" </span></span>
<span><span class="co">##                         cell_type_coarse </span></span>
<span><span class="co">##                                   "NULL" </span></span>
<span><span class="co">##                           cell_type_fine </span></span>
<span><span class="co">##                                   "NULL" </span></span>
<span><span class="co">##                        cell_type_general </span></span>
<span><span class="co">##                                   "NULL" </span></span>
<span><span class="co">##                            cell_type_med </span></span>
<span><span class="co">##                                   "NULL" </span></span>
<span><span class="co">##                                    H_knn </span></span>
<span><span class="co">##                                   "NULL" </span></span>
<span><span class="co">##                                mito_frac </span></span>
<span><span class="co">##                                "numeric" </span></span>
<span><span class="co">##                                 RBP_frac </span></span>
<span><span class="co">##                                "numeric"</span></span></code></pre>
<p>This example will identify relevant metadata following the guiding
questions below. The questions apply to other objects as well.</p>
<div class="card mb-3">
<div class="card-header">
<h5 class="card-title center">
Guiding Questions for Metadata Selection
</h5>
</div>
<div class="card-body">
<ul>
<li>Which variables represent cell type annotation, or clustering?</li>
<li>If there are multiple variables for cell type, which are most
relevant to the end user?</li>
<li>Do the different variables reflect resolutions for cell type, and if
so, which resolution is most appropriate?</li>
<li>Is there a variable for patient (or sample)?</li>
<li>Are there clinically relevant variables, or variables reflecting
research goals of the study?</li>
<li>Are there any other variables that delineate subsets of cells users
may find interesting?</li>
</ul>
</div>
</div>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>While Numeric metadata can’t be used for subsetting or choices on
plots, it may be plotted like features, if allowed via the config app.
See the config app section for instructions on how to enable/disable
this usage.</p>
</div>
<p>In the Chan et al. dataset, there are six metadata variables related
to cell type: cell_type_coarse, cell_type, cell_type_fine,
cell_type_general, clusters, and cell_type_med.</p>
<p>To identify which are relevant, one approach is to look at the unique
values for each variable. You can get this information in a single
function call using the <code>unique_values()</code> function from the
<a href="https://github.com/amc-heme/SCUBA#readme" class="external-link">SCUBA</a>
package.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/unique_values.html" class="external-link">unique_values</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">full_dataset</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"cell_type_fine"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] T cell         SCLC-N         Endothelial    SCLC-A         SCLC-P        </span></span>
<span><span class="co">##  [6] AE1            Basal          Macrophage     DC             NSCLC         </span></span>
<span><span class="co">## [11] Ciliated       Neuroendocrine Plasma cell    Ionocyte       Hepatocyte    </span></span>
<span><span class="co">## [16] Mast           B cell         Fibroblast     AEP            Neutrophil    </span></span>
<span><span class="co">## [21] Mucinous       Erythrocyte    Club           Tuft          </span></span>
<span><span class="co">## 24 Levels: AE1 AEP B cell Basal Ciliated Club DC Endothelial ... Tuft</span></span></code></pre>
<p>It may also be helpful to look at the number of unique values in the
object for each variable.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_type_vars</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type_coarse"</span>, </span>
<span>    <span class="st">"cell_type_fine"</span>, </span>
<span>    <span class="st">"cell_type_general"</span>, </span>
<span>    <span class="st">"cell_type"</span>,</span>
<span>    <span class="st">"clusters"</span>, </span>
<span>    <span class="st">"cell_type_med"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">meta_var</span> <span class="kw">in</span> <span class="va">cell_type_vars</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">meta_var</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Pass results of unique_values to length() to get the number of values</span></span>
<span>  <span class="va">n_unique</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/unique_values.html" class="external-link">unique_values</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">full_dataset</span>,</span>
<span>    var <span class="op">=</span> <span class="va">meta_var</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">n_unique</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "cell_type_coarse"</span></span>
<span><span class="co">## [1] 4</span></span>
<span><span class="co">## [1] "cell_type_fine"</span></span>
<span><span class="co">## [1] 24</span></span>
<span><span class="co">## [1] "cell_type_general"</span></span>
<span><span class="co">## [1] 3</span></span>
<span><span class="co">## [1] "cell_type"</span></span>
<span><span class="co">## [1] 11</span></span>
<span><span class="co">## [1] "clusters"</span></span>
<span><span class="co">## [1] 56</span></span>
<span><span class="co">## [1] "cell_type_med"</span></span>
<span><span class="co">## [1] 16</span></span></code></pre>
<p>We see that <code>cell_type_general</code> and
<code>cell_type_coarse</code> have the lowest resolution, and that there
are very many clusters in <code>clusters</code>.
<code>cell_type_fine</code> and <code>cell_type_med</code> have a
middling resolution.</p>
<p>Which resolution is most appropriate? Let’s plot the each one find
out.
<!-- Let's plot each metadata variable to see which resolution is most appropriate. --></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">cell_type_vars</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type_general"</span>, </span>
<span>    <span class="st">"cell_type_coarse"</span>,</span>
<span>    <span class="st">"cell_type"</span>,</span>
<span>    <span class="st">"cell_type_med"</span>, </span>
<span>    <span class="st">"cell_type_fine"</span>, </span>
<span>    <span class="st">"clusters"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">meta_var</span> <span class="kw">in</span> <span class="va">cell_type_vars</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">plots</span><span class="op">[[</span><span class="va">meta_var</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/plot_reduction.html" class="external-link">plot_reduction</a></span><span class="op">(</span></span>
<span>      <span class="va">full_dataset</span>,</span>
<span>      group_by <span class="op">=</span> <span class="va">meta_var</span>,</span>
<span>      label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>      <span class="op">)</span> </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="dataset_setup_walkthrough-output_full_object_plots.png"></p>
<p>From the plots, we see that the resolution in
<code>cell_type_coarse</code> matches the resolution in <strong>Figure
1A</strong>.</p>
<p>Next, we look at patient or sample metadata. There are also multiple
variables for sample metadata in this object: <code>donor_id</code>,
<code>HTAN_Biospecimen_ID</code>, <code>HTAN_Participant_ID</code>, and
<code>batch</code>. The levels in the <code>donor_id</code> variable
most closely match the patient IDs specified by Chan et al. 2021 in
<strong>Supplementary Table S1</strong>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/unique_values.html" class="external-link">unique_values</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">full_dataset</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"donor_id"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_order.html" class="external-link">str_sort</a></span><span class="op">(</span>numeric <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre class="scroll-100"><code><span><span class="co">##  [1] "PleuralEffusion" "RU255"           "RU426"           "RU653"          </span></span>
<span><span class="co">##  [5] "RU661"           "RU666"           "RU675"           "RU676"          </span></span>
<span><span class="co">##  [9] "RU681"           "RU682"           "RU684"           "RU685"          </span></span>
<span><span class="co">## [13] "RU699"           "RU701"           "RU779"           "RU1038"         </span></span>
<span><span class="co">## [17] "RU1057"          "RU1061"          "RU1065"          "RU1066"         </span></span>
<span><span class="co">## [21] "RU1080"          "RU1108"          "RU1124"          "RU1128"         </span></span>
<span><span class="co">## [25] "RU1134"          "RU1135"          "RU1137"          "RU1138"         </span></span>
<span><span class="co">## [29] "RU1144"          "RU1145"          "RU1152"          "RU1170"         </span></span>
<span><span class="co">## [33] "RU1181"          "RU1195"          "RU1215"          "RU1229"         </span></span>
<span><span class="co">## [37] "RU1231"          "RU1262"          "RU1271"          "RU1293"         </span></span>
<span><span class="co">## [41] "RU1311"          "RU1322"</span></span></code></pre>
<p><code>treatment</code>, <code>tissue</code>, and
<code>self_reported_ethnicity</code> are also of interest, since they
allow for comparison of cells in the app based on clinical metadata.
<code>treatment</code> and <code>tissue</code> appear in <strong>Figure
1B</strong> and <strong>Figure 1C</strong> of Chan et al.,
respectively.</p>
</div>
<div class="section level3">
<h3 id="modalities-and-reductions">Modalities and Reductions<a class="anchor" aria-label="anchor" href="#modalities-and-reductions"></a>
</h3>
<p>The features included in the app for plotting are based on the
included modalities. For anndata objects, expression data for features
in the primary modality is stored in the <code>X</code> slot, and data
for additional modalities is stored in <code>obsm</code>. Dimensional
reduction coordinates for each cell are also stored in
<code>obsm</code>. Let’s use the <code>obsm_keys()</code> method to see
what alternate modalities and reductions exist.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_dataset</span><span class="op">$</span><span class="fu">obsm_keys</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "X_pca"  "X_umap"</span></span></code></pre>
<p>We observe see two matrices in <code>obsm</code>: <code>X_umap</code>
and <code>X_pca</code>. These correspond to the UMAP and PCA reductions,
respectively.</p>
<p>The UMAP reduction is preferred as it is the current standard for
scRNA-seq visualization. The PCA reduction is of limited utility,
partially due to the number of dimensions. Plotting in the app is best
suited for reductions with exactly two dimensions, so only the UMAP
reduction should be included.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Number of dimensions in each reduction</span></span>
<span><span class="va">full_dataset</span><span class="op">$</span><span class="va">obsm</span><span class="op">$</span><span class="va">X_umap</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 147137      2</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_dataset</span><span class="op">$</span><span class="va">obsm</span><span class="op">$</span><span class="va">X_pca</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 147137     50</span></span></code></pre>
<p>No alternate modalities are observed in the <code>obsm</code> slot
for this object. There is only the main modality, gene expression, in
the <code>X</code> slot.</p>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>Other datasets may include surface protein (ADT) data, computed gene
signatures per cell, or other data. <strong>For anndata objects,
alternate modalities must be stored in <code>obsm</code> to be
accessible to scExploreR</strong> (with the exception of <code>X</code>,
of course). Seurat and SingleCellExperiment objects have defined slots
for modalities: <code>@assays</code> in Seurat objects, and “alternate
experiments” for SingleCellExperiment objects. For these object types,
scExploreR will use these slots to access modality information. As long
as feature expression data for each modality is stored using the
convention for Seurat and SingleCellExperiment objects, no additional
action is needed for scExploreR to find this data.</p>
</div>
</div>
<div class="section level3">
<h3 id="sclc-epithelial-cells-object">SCLC Epithelial Cells Object<a class="anchor" aria-label="anchor" href="#sclc-epithelial-cells-object"></a>
</h3>
<p>Next, we explore the SCLC Epithelial Cells dataset to see if the
characteristics exposed differ from the combined dataset.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://anndata.dynverse.org" class="external-link">anndata</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">sclc_dataset</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">anndata</span><span class="fu">::</span><span class="fu"><a href="https://anndata.dynverse.org/reference/read_h5ad.html" class="external-link">read_h5ad</a></span><span class="op">(</span></span>
<span>    <span class="co"># Replace the code below with the path to the downloaded dataset</span></span>
<span>    <span class="st">"~/datasets/small_cell_lung_cancer/Epithelial_SCLC.h5ad"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<!-- Figure out the 'imputed_normalized', 'log2(X+0.1)', 'normalized' layers -->
<p>To determine if there are metadata variables that are in the SCLC
object that are not in the full object, run <code><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff()</a></code> on
the variable names from each. If you want to see variables that are in
the full object that are not in the SCLC object, switch the order of the
parameters to <code><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff()</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify metadata variables that are specific to the SCLC object</span></span>
<span><span class="va">sclc_unique_vars</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span></span>
<span>    <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/meta_varnames.html" class="external-link">meta_varnames</a></span><span class="op">(</span><span class="va">sclc_dataset</span><span class="op">)</span>,</span>
<span>    <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/meta_varnames.html" class="external-link">meta_varnames</a></span><span class="op">(</span><span class="va">full_dataset</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">sclc_unique_vars</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "n_genes_by_counts"             "log1p_n_genes_by_counts"      </span></span>
<span><span class="co">##  [3] "total_counts"                  "log1p_total_counts"           </span></span>
<span><span class="co">##  [5] "SCLC_subtype"                  "SCLC_subtype_plus_TP53_RB1_wt"</span></span>
<span><span class="co">##  [7] "pval_SCLC-A"                   "pval_SCLC-N"                  </span></span>
<span><span class="co">##  [9] "pval_SCLC-P"                   "NEUROD genes"                 </span></span>
<span><span class="co">## [11] "MHCI_PRESENTATION"             "MHCII_PRESENTATION"           </span></span>
<span><span class="co">## [13] "SCLC_major_subtype_of_sample"  "treatment_categorized"        </span></span>
<span><span class="co">## [15] "recurrent_cluster"             "clusters_fine"</span></span></code></pre>
<!-- Don't forget to mention in the full object section that only categorical            metadata can be added to scExploreR -->
<p>From this we see 16 variables that are not in the full object. Let’s
view which variables are categorical.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Iterate through metadata variables and return a vector specifying whether </span></span>
<span><span class="co"># the variable is categorical (character, factor), or numeric (numeric, </span></span>
<span><span class="co"># double, integer)</span></span>
<span><span class="va">var_classes</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span></span>
<span>    <span class="va">sclc_unique_vars</span>,</span>
<span>    <span class="co"># Compute the class of the variable</span></span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">meta_var</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sclc_dataset</span><span class="op">$</span><span class="va">obs</span><span class="op">[[</span><span class="va">meta_var</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Sort to group classes together</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">var_classes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  SCLC_subtype SCLC_subtype_plus_TP53_RB1_wt </span></span>
<span><span class="co">##                      "factor"                      "factor" </span></span>
<span><span class="co">##  SCLC_major_subtype_of_sample         treatment_categorized </span></span>
<span><span class="co">##                      "factor"                      "factor" </span></span>
<span><span class="co">##             recurrent_cluster                 clusters_fine </span></span>
<span><span class="co">##                      "factor"                      "factor" </span></span>
<span><span class="co">##             n_genes_by_counts       log1p_n_genes_by_counts </span></span>
<span><span class="co">##                     "numeric"                     "numeric" </span></span>
<span><span class="co">##                  total_counts            log1p_total_counts </span></span>
<span><span class="co">##                     "numeric"                     "numeric" </span></span>
<span><span class="co">##                   pval_SCLC-A                   pval_SCLC-N </span></span>
<span><span class="co">##                     "numeric"                     "numeric" </span></span>
<span><span class="co">##                   pval_SCLC-P                  NEUROD genes </span></span>
<span><span class="co">##                     "numeric"                     "numeric" </span></span>
<span><span class="co">##             MHCI_PRESENTATION            MHCII_PRESENTATION </span></span>
<span><span class="co">##                     "numeric"                     "numeric"</span></span></code></pre>
<p>We see that SCLC_subtype, SCLC_subtype_plus_TP53_RB1_wt,
SCLC_major_subtype_of_sample, treatment_categorized, recurrent_cluster,
and clusters_fine are categorical variables. Let’s plot them below to
better understand them. <code>clusters</code> and <code>treatment</code>
are added for comparison with <code>clusters_fine</code> and
<code>treatment_categorized</code>, respectively.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_type_vars</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SCLC_subtype"</span>,</span>
<span>    <span class="st">"SCLC_subtype_plus_TP53_RB1_wt"</span>,</span>
<span>    <span class="st">"SCLC_major_subtype_of_sample"</span>,</span>
<span>    <span class="st">"treatment_categorized"</span>,</span>
<span>    <span class="st">"treatment"</span>,</span>
<span>    <span class="st">"recurrent_cluster"</span>,</span>
<span>    <span class="st">"clusters_fine"</span>,</span>
<span>    <span class="st">"clusters"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">meta_var</span> <span class="kw">in</span> <span class="va">cell_type_vars</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">plots</span><span class="op">[[</span><span class="va">meta_var</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/plot_reduction.html" class="external-link">plot_reduction</a></span><span class="op">(</span></span>
<span>      <span class="va">sclc_dataset</span>,</span>
<span>      group_by <span class="op">=</span> <span class="va">meta_var</span>,</span>
<span>      label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>      <span class="op">)</span> </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="dataset_setup_walkthrough-output_sclc_categorical_plots.png"></p>
<p>From the plots, we see that <code>SCLC_major_subtype_of_sample</code>
is identical to <code>SCLC_subtype</code>, and
<code>SCLC_subtype_plus_TP53_RB1_wt</code> has the the same information
as the former variables, but includes “TP53/RB1-wt SCLC”. <strong>Figure
1G</strong> of Chan et al. Includes this designation, so we will use
<code>SCLC_subtype_plus_TP53_RB1_wt</code>.</p>
<p>For clustering, the clusters in <code>clusters</code> appear to align
with the representation of cluster 22 in <strong>Figure 2C</strong>, and
<code>clusters_fine</code> does not, so we will use
<code>clusters</code>.</p>
<p>The <code>recurrent_cluster</code> variable appears to align with
cluster 22 in <code>clusters</code>. Below we test if the cell IDs in
cluster 22 are identical to the IDs for cells where
<code>recurrent_cluster</code> is TRUE.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use SCUBA::get_all_cells to return cell IDs for cells with a value </span></span>
<span><span class="co"># of TRUE for the "recurrent cluster" column</span></span>
<span><span class="va">recurrent_cluster_cell_names</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/get_all_cells.html" class="external-link">get_all_cells</a></span><span class="op">(</span></span>
<span>    <span class="co"># All cells from the subset below will be returned</span></span>
<span>    <span class="va">sclc_dataset</span><span class="op">[</span><span class="va">sclc_dataset</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">recurrent_cluster</span> <span class="op">==</span> <span class="st">"True"</span>, <span class="op">]</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Return cell IDs for cluster 22</span></span>
<span><span class="va">cluster_22_cell_names</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">SCUBA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SCUBA/man/get_all_cells.html" class="external-link">get_all_cells</a></span><span class="op">(</span></span>
<span>    <span class="va">sclc_dataset</span><span class="op">[</span><span class="va">sclc_dataset</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">"22"</span>, <span class="op">]</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Test if ID vectors are identical</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span></span>
<span>  <span class="va">recurrent_cluster_cell_names</span>,</span>
<span>  <span class="va">cluster_22_cell_names</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>The cells where <code>recurrent_cluster</code> is TRUE are the same
cells as those in cluster 22. Since cluster 22 recurs across patients
(<strong>Figures 3 A-C</strong> of Chan et al.), and is shown to be
associated with a malignant phenotype (<strong>Figure S3B</strong>),
including a binary designation for this cluster would create a useful
shortcut for end users wishing to examine cells in cluster 22 versus
other cells. We will therefore include this variable for the SCLC
object.</p>
<p>For <code>treatment_categorized</code>, the values in this variable
align with the information in <strong>Figure 1B</strong> and
<strong>Table S1</strong> of Chan et al. We will use
<code>treatment_categorized</code> instead of
<code>treatment</code>.</p>
<div class="section level4">
<h4 id="modalities-and-reductions-1">Modalities and Reductions<a class="anchor" aria-label="anchor" href="#modalities-and-reductions-1"></a>
</h4>
<p>There are no alternate modalities in <code>obsm</code>. The
reductions in <code>obsm</code> are the same as for the main object,
with the addition of <code>X_diffmap</code>. This reduction does not
appear in the figures of Chan et al. and is not explicitly mentioned in
the methods section, so it is best to exclude it to keep things simple
for the end user.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="configure-dataset-specific-scexplorer-settings-in-the-config-app">Configure Dataset-specific scExploreR Settings in the Config
App<a class="anchor" aria-label="anchor" href="#configure-dataset-specific-scexplorer-settings-in-the-config-app"></a>
</h2>
<p>Now that we have identified metadata, reductions, and modalities to
include for the dataset, the next step is to apply these settings in the
config app.</p>
<div class="section level3">
<h3 id="anndata-specific-preparation-for-the-config-app">Anndata-specific Preparation for the Config App<a class="anchor" aria-label="anchor" href="#anndata-specific-preparation-for-the-config-app"></a>
</h3>
<p>Before loading Anndata objects into the config app, the keys of
alternate modalities and reductions must be defined, since both types of
data are stored in <code>obsm</code>, and there is no exclusive storage
slot for reductions or modalities in anndata objects.</p>
<p>Alternate modalities and reductions should be stored in
<code>uns</code>, under <code>scExploreR_assays</code> and
<code>scExploreR_reductions</code>, respectively. This is done below in
python, via reticulate, for both the full object and the SCLC
object.</p>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>This step is not required for Seurat or SingleCellExperiment
objects.</p>
</div>
<div class="sourceCode" id="cb27"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="co"># If you are working in RStudio, make sure to replace the `r` in the </span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a><span class="co"># code chunk with `python`</span></span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a><span class="co"># Full object</span></span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a><span class="co"># Use `r.` to access the full dataset in python using Reticulate</span></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a><span class="co"># No non-RNA modalities</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>r.full_dataset.uns[<span class="st">"scExploreR_assays"</span>] <span class="op">=</span> []</span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>r.full_dataset.uns[<span class="st">"scExploreR_reductions"</span>] <span class="op">=</span> [<span class="st">"X_umap"</span>, <span class="st">"X_pca"</span>]</span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a><span class="co"># SCLC object</span></span>
<span id="cb27-12"><a href="#cb27-12" tabindex="-1"></a>r.sclc_dataset.uns[<span class="st">"scExploreR_assays"</span>] <span class="op">=</span> []</span>
<span id="cb27-13"><a href="#cb27-13" tabindex="-1"></a></span>
<span id="cb27-14"><a href="#cb27-14" tabindex="-1"></a><span class="co"># Indicate that X_diffmap is a reduction, even though it will not be used</span></span>
<span id="cb27-15"><a href="#cb27-15" tabindex="-1"></a>r.sclc_dataset.uns[<span class="st">"scExploreR_reductions"</span>] <span class="op">=</span> [<span class="st">"X_umap"</span>, <span class="st">"X_pca"</span>, <span class="st">"X_diffmap"</span>]</span></code></pre></div>
<p>After adding the data to <code>uns</code>, make sure to save both
objects. They can now be loaded into the config app for analysis.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">anndata</span><span class="fu">::</span><span class="fu"><a href="https://anndata.dynverse.org/reference/write_h5ad.html" class="external-link">write_h5ad</a></span><span class="op">(</span></span>
<span>  <span class="va">full_dataset</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"~/datasets/small_cell_lung_cancer/SCLC_Full.h5ad"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">anndata</span><span class="fu">::</span><span class="fu"><a href="https://anndata.dynverse.org/reference/write_h5ad.html" class="external-link">write_h5ad</a></span><span class="op">(</span></span>
<span>  <span class="va">sclc_dataset</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"~/datasets/small_cell_lung_cancer/Epithelial_SCLC.h5ad"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="launching-the-config-app">Launching the Config App<a class="anchor" aria-label="anchor" href="#launching-the-config-app"></a>
</h3>
<p>To launch the config app, run the code below replacing
<code>object_path</code> with the path to the downloaded dataset.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scExploreR</span><span class="fu">::</span><span class="fu"><a href="../reference/run_config.html">run_config</a></span><span class="op">(</span></span>
<span>  object_path <span class="op">=</span> <span class="st">"~/datasets/small_cell_lung_cancer/SCLC_Full.h5ad"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Upon loading the config app, you will notice several tabs. The
general tab is selected by default. This tab is used to enter basic
information on the dataset, which is displayed in the app on the menu
that appears when switching between datasets. This is useful when an app
is created with multiple datasets, which is the case here.</p>
<p><img src="dataset_setup_walkthrough-001-tabs_and_general_dataset_window.png"></p>
<p>Use the <strong>“Label for Dataset”</strong> field to add a title for
the dataset. In this case, we will use “Small Cell Lung Carcinoma
Atlas”.</p>
<p><img src="dataset_setup_walkthrough-003-dataset-label.png"></p>
<p>In the <strong>“Description of Dataset”</strong> field, add a summary
of the dataset. This should be a high-level description, which can
include information like the number of cells, the number of samples, the
types of samples included, and other relevant details on the
experimental design.</p>
<div class="accordion autoclose info" data-require-bs-caller="accordion()" data-require-bs-version="5" id="bslib-accordion-6363">
<div class="accordion-item" data-value="Suggested description for this example">
<div class="accordion-header">
<button class="accordion-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#bslib-accordion-panel-5946" aria-expanded="false" aria-controls="bslib-accordion-panel-5946">
<div class="accordion-icon"></div>
<div class="accordion-title">Suggested description for this example</div>
</button>
</div>
<div id="bslib-accordion-panel-5946" class="accordion-collapse collapse" data-bs-parent="#bslib-accordion-6363">
<div class="accordion-body" style="margin-bottom: 15px;">
<p>147k cells from 19 patients. Twenty-one small cell lung cancer samples are included, along with 24 lung
      adenocarcinoma samples, and four tumor-adjacent normal lung samples included as controls. See Chan et al. 
      Cancer Cell 2021 for more details.</p>
</div>
</div>
</div>
</div>
<p><img src="dataset_setup_walkthrough-004-dataset-description.png"></p>
<p>Next, use the <strong>“Content for Dataset Preview”</strong> menu to
add a plot thumbnail used to preview the dataset in the selection menu.
In this example, the plot will serves as an additional guide when
switching between the datasets, since they can be compared against the
figures in the paper. You can use the settings below to determine which
metadata to show, and which reduction to use (in this example, this
should be “X_umap”).</p>
<p><img src="dataset_setup_walkthrough-006-preview_dimplot.png"></p>
<div class="section level4">
<h4 id="assays-modalities-tab">Assays (Modalities) Tab<a class="anchor" aria-label="anchor" href="#assays-modalities-tab"></a>
</h4>
<p>Switch to the assays tab to enter information about the modalities in
the object. We identified previously that there is only one modality in
this object (gene expression, in <code>X</code>).</p>
<p><img src="dataset_setup_walkthrough-007-assay_tab_intro.png"></p>
<p>To include a modality in scExploreR, select it under
<strong>“Available Assays”</strong>. Once selected, it will appear under
<strong>“Selected Assays”</strong>, and a card will appear on the right
side of the screen with modality-specific options. Use the text entry to
provide a display name for the modality. The display name should be
singular instead of plural since the name will display in the context of
a single feature from that assay, so in this case, it should be
“Gene”.</p>
<p>If you want the name to display in plots, click the checkbox. This is
helpful when you have features in multiple modalities with the same
name, though in this example there is only one modality, so the box can
be left unchecked.</p>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>When configuring an object with multiple modalities, the checkbox for
the “Gene” assay should generally not be checked. Usually most of the
features in an object will be genes, so it is not necessary to specify
that a feature is a gene. For other modalities, it generally does make
sense to add the name of modality to the plot. It is especially helpful
to add a label when feature names overlap with names in the gene assay.
This can be the case for ADT data, where several surface proteins
overlap with human gene symbols (i.e. CD34 and CD36).</p>
</div>
<p><img src="dataset_setup_walkthrough-008-assay_tab_select_X_and_options.png"></p>
</div>
<div class="section level4">
<h4 id="metadata-tab">Metadata Tab<a class="anchor" aria-label="anchor" href="#metadata-tab"></a>
</h4>
<p>The metadata tab is used to identify the metadata variables to use in
the app, and specify settings for each variable.</p>
<p><img src="dataset_setup_walkthrough-010-metadata_introduction.png"></p>
<p>Drag a metadata variable from the <strong>“Available
Metadata”</strong> column to the <strong>“Selected Metadata”</strong>
column to include it in the app. Use the text box to enter a display
name for the variable. Drag <strong>“clusters”</strong> to the
<strong>“Selected Metadata”</strong> column and add the label
<strong>“Clusters”</strong>.</p>
<p><img src="dataset_setup_walkthrough-011-Drag_Metadata_Demo.png"></p>
<p>You can drag the variables in the <strong>“Selected
Metadata”</strong> column to change their order. Menus in the app will
reflect the order of the variables.</p>
<p><img src="dataset_setup_walkthrough-012-Reorder_Metadata.png"></p>
<p>Drag the metadata columns listed below to the <strong>“Selected
Metadata”</strong> column, and add the suggested display names.</p>
<div class="card mb-3">
<div class="card-header">
<h5 class="card-title center">
Suggested Metadata Options
</h5>
</div>
<div class="card-body">
<div class="datatables html-widget html-fill-item" id="htmlwidget-a734f6328c7ad284b53d" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-a734f6328c7ad284b53d">{"x":{"style":"bootstrap5","filter":"none","vertical":false,"data":[["clusters","cell_type_fine","treatment","tissue","donor_id","HTAN_Biospecimen_ID","self_reported_ethnicity"],["Clusters","Cell Type","Treatment History","Tissue","Patient ID","HTAN Sample ID","Self-Reported Ethnicity"]],"container":"<table class=\"table table-striped table-hover row-border order-column display\">\n  <thead>\n    <tr>\n      <th><b>Selected Variable<\/b><\/th>\n      <th><b>Label<\/b><\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"searching":false,"paging":false,"info":false,"ordering":false,"columnDefs":[{"name":"Selected.Variable","targets":0},{"name":"Label","targets":1}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div class="section level4">
<h4 id="reductions-tab">Reductions Tab<a class="anchor" aria-label="anchor" href="#reductions-tab"></a>
</h4>
<p>Next, set settings for reductions in the reductions tab. All
reductions are included by default, though the PCA in this example is
not useful for plotting, so there is no reason to include it. Select the
PCA reduction to remove it.</p>
<p><img src="dataset_setup_walkthrough-013-reductions_intro.png"></p>
<p><img src="dataset_setup_walkthrough-014-remove_PCA.png"></p>
<p>The labels for each reduction default to the keys assigned to them in
the object. Let’s rename the label for the UMAP reduction to “UMAP”.</p>
<p><img src="dataset_setup_walkthrough-015-change_reduction_label.png"></p>
</div>
<div class="section level4">
<h4 id="save-config-file">Save Config file<a class="anchor" aria-label="anchor" href="#save-config-file"></a>
</h4>
<p>Now that we’ve filled out all the information in the config file,
let’s save the file to disk. To do this, navigate to the orange
<i class="fas fa-ellipsis" role="presentation" aria-label="ellipsis icon"></i> button, select <strong>“Save Config
File”</strong>, and use the system prompt that appears to save the file.
The file will save with a “.yaml” extension.</p>
<p><img src="dataset_setup_walkthrough-016-config_save_file.png"></p>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>The recommended filename is the the name of your object plus
“_config”, though you can use any name you wish.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="create-config-file-for-epithelial-sclc-object">Create Config File for Epithelial SCLC Object<a class="anchor" aria-label="anchor" href="#create-config-file-for-epithelial-sclc-object"></a>
</h3>
<p>Next, we’ll create the config file for the SCLC object. Run the
config app and repeat the process above, using the suggested settings
below as a guide.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scExploreR</span><span class="fu">::</span><span class="fu"><a href="../reference/run_config.html">run_config</a></span><span class="op">(</span></span>
<span>  object_path <span class="op">=</span> <span class="st">"~/datasets/small_cell_lung_cancer/Epithelial_SCLC.h5ad"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="card mb-3">
<div class="card-header">
<h5 class="card-title center">
Suggested Settings for Epithelial SCLC Object
</h5>
</div>
<div class="card-body">
<h5>
<strong>General Tab</strong>
</h5>
<p><strong>Label for dataset:</strong> SCLC Epithelial Cells</p>
<p><strong>Description:</strong> 54k cells from epithelial tissue,
designated as SCLC tumor cells. This is a subset of the combined samples
dataset that has been re-clustered. For more information, see Chan et
al. Cancer Cell 2021.</p>
<p><strong>Dataset Preview:</strong></p>
<ul>
<li>DimPlot</li>
<li>Metadata to Group By: SCLC_subtype_plus_TP53_RB1_wt</li>
<li>Metadata to Split By: none</li>
<li>Choose Reduction: X_umap</li>
<li>Label Groups: Checked</li>
<li>Show Legend: Checked</li>
</ul>
<div class="section level5">
<h5 id="assays-tab">
<strong>Assays Tab</strong><a class="anchor" aria-label="anchor" href="#assays-tab"></a>
</h5>
<p><strong>Assays selected:</strong> X</p>
<p><strong>Options for X</strong></p>
<ul>
<li>
<strong>Label:</strong> “Gene”</li>
<li>
<strong>Checkbox:</strong> unchecked</li>
</ul>
</div>
<div class="section level5">
<h5 id="metadata-tab-1">
<strong>Metadata Tab</strong><a class="anchor" aria-label="anchor" href="#metadata-tab-1"></a>
</h5>
<p>Table of selected variables, with their corresponding labels. The
suggested order is the order of the variables on the table.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-df174099310d81fbbd65" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-df174099310d81fbbd65">{"x":{"style":"bootstrap5","filter":"none","vertical":false,"data":[["SCLC_subtype","treatment_categorized","donor_id","tissue","clusters","recurrent_cluster","self_reported_ethnicity"],["SCLC Subtype","Treatment","Patient ID","Tissue","Clusters","Recurrent Cluster","Self-Reported Ethnicity"]],"container":"<table class=\"table table-striped table-hover row-border order-column display\">\n  <thead>\n    <tr>\n      <th><b>Selected Variable<\/b><\/th>\n      <th><b>Label<\/b><\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"searching":false,"paging":false,"info":false,"ordering":false,"columnDefs":[{"name":"Selected.Variable","targets":0},{"name":"Label","targets":1}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level5">
<h5 id="reductions-tab-1">
<strong>Reductions Tab</strong><a class="anchor" aria-label="anchor" href="#reductions-tab-1"></a>
</h5>
<p>Reductions selected: X_umap</p>
<p>Options for X_umap - Label: “UMAP”</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="create-an-app-config-file">Create an App Config File<a class="anchor" aria-label="anchor" href="#create-an-app-config-file"></a>
</h3>
<p>The final step before launching scExploreR is to create an app config
file. The app config file specifies the paths of all objects used in the
app and their respective config files, and it also defines app-wide
settings. An example app config file is given below.</p>
<p><code>datasets</code> is a list of all the datasets in the app. Each
key under <code>datasets</code> represents a dataset to be loaded in the
app instance (in this case, the dataset keys are
<code>combined_sclc</code> and <code>epithelial_sclc</code>). The keys
under <code>datasets</code> can be any name, as long as spaces are not
used (use underscores instead). Any descriptive name for the dataset
will work.</p>
<p>Under the dataset keys, there are two keys: <code>object</code> and
<code>config</code>. These should be the path to the object and config
file, respectively.</p>
<p>The <code>RSTUDIO_PANDOC</code> key is used to define a path to an
RStudio pandoc installation. This should only be necessary when running
on a linux server.</p>
<p>To determine if this is necessary, run
<code><a href="https://pkgs.rstudio.com/rmarkdown/reference/pandoc_available.html" class="external-link">rmarkdown::pandoc_available()</a></code> in your R console. If
<code>TRUE</code> is returned, <code>RSTUDIO_PANDOC</code> does not need
to be specified. In place of a path, enter a YAML <code>NULL</code>
(<code>~</code>) for this key. If you see <code>FALSE</code>, you will
need to install pandoc manually and provide the path to the
installation. <code><a href="https://pkgs.rstudio.com/rmarkdown/reference/find_pandoc.html" class="external-link">rmarkdown::find_pandoc()</a></code> may be able to find
the path.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="co"># YAML version</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="pp">%YAML 1.1</span></span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a><span class="pp">---</span></span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a><span class="fu">datasets</span><span class="kw">:</span></span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a><span class="at">    </span><span class="fu">combined_sclc</span><span class="kw">:</span></span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a><span class="co">        # Replace path with path to object</span></span>
<span id="cb31-7"><a href="#cb31-7" tabindex="-1"></a><span class="at">        </span><span class="fu">object</span><span class="kw">:</span><span class="at"> </span><span class="st">"~/datasets/small_cell_lung_cancer/SCLC_Full.h5ad"</span></span>
<span id="cb31-8"><a href="#cb31-8" tabindex="-1"></a><span class="co">        # Replace with path to config file</span></span>
<span id="cb31-9"><a href="#cb31-9" tabindex="-1"></a><span class="at">        </span><span class="fu">config</span><span class="kw">:</span><span class="at"> </span><span class="st">"~/datasets/small_cell_lung_cancer/SCLC_Full_config.yaml"</span></span>
<span id="cb31-10"><a href="#cb31-10" tabindex="-1"></a><span class="at">    </span><span class="fu">epithelial_sclc</span><span class="kw">:</span></span>
<span id="cb31-11"><a href="#cb31-11" tabindex="-1"></a><span class="at">        </span><span class="fu">object</span><span class="kw">:</span><span class="at"> </span><span class="st">"~/datasets/small_cell_lung_cancer/Epithelial_SCLC.h5ad"</span></span>
<span id="cb31-12"><a href="#cb31-12" tabindex="-1"></a><span class="at">        </span><span class="fu">config</span><span class="kw">:</span><span class="at"> </span><span class="st">"~/datasets/small_cell_lung_cancer/Epithelial_SCLC_config.yaml"</span></span>
<span id="cb31-13"><a href="#cb31-13" tabindex="-1"></a><span class="fu">RSTUDIO_PANDOC</span><span class="kw">:</span><span class="at"> </span><span class="ch">~</span></span></code></pre></div>
<p>You can create a config file from R Studio by going to
<strong>File</strong> &gt;&gt; <strong>New File</strong> &gt;&gt;
<strong>Text File</strong>. After creating the text file, save it in the
same directory as the object and config files, and <strong>be sure to
add “.yaml” to the filename</strong>. After saving initially, use the
structure above and define the paths to the object and config files.</p>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>Don’t forget the YAML version at the top. Version 1.1 is the latest
version supported by the <code>yaml</code> package used to read the data
into R.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="run-scexplorer">Run scExploreR<a class="anchor" aria-label="anchor" href="#run-scexplorer"></a>
</h2>
<p>The app can now be ran by providing the path to the app config file.
Use the <code>browser_config</code> parameter to specify the path.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scExploreR</span><span class="fu">::</span><span class="fu"><a href="../reference/run_scExploreR.html">run_scExploreR</a></span><span class="op">(</span></span>
<span>  browser_config <span class="op">=</span> <span class="st">"~/datasets/small_cell_lung_cancer/sclc_app_config.yaml"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="callout callout-info">
<h5>
Note
</h5>
<p>If you only have one object, you don’t need to make an app config
file. You can simply run <code>run_scExploreR</code>, using the
<code>object_path</code> and <code>config_path</code> parameters instead
of <code>browser_config</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scExploreR</span><span class="fu">::</span><span class="fu"><a href="../reference/run_scExploreR.html">run_scExploreR</a></span><span class="op">(</span></span>
<span>  object_path <span class="op">=</span> <span class="st">"~/datasets/small_cell_lung_cancer/SCLC_Full.h5ad"</span>,</span>
<span>  config_path <span class="op">=</span> <span class="st">"~/datasets/small_cell_lung_cancer/SCLC_Full_config.yaml"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Hematology Informatics Group.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
