<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scExploreR Documentation • scExploreR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scExploreR Documentation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scExploreR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Help and Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/dataset_setup_walkthrough.html">App Setup Walkthrough</a></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">Exploring your Data</a></li>
    <li><a class="dropdown-item" href="../articles/object_guidelines.html">Object Preparation Guidelines</a></li>
    <li><a class="dropdown-item" href="../articles/docker.html">Deploying With Docker</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Additional Information</h6></li>
    <li><a class="dropdown-item" href="../articles/scRNA_Plots_Explained.html">Guide to scRNA-Seq Plots</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_subsetting_documentation.html">String Subsetting Guide</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Full Documentation</h6></li>
    <li><a class="dropdown-item" href="../articles/full_documentation.html">scExploreR Documentation</a></li>
    <li><a class="dropdown-item" href="../articles/config_documentation.html">Config App Documentation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/amc-heme/scExploreR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="full_documentation_files/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="full_documentation_files/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scExploreR Documentation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amc-heme/scExploreR/blob/main/vignettes/full_documentation.Rmd" class="external-link"><code>vignettes/full_documentation.Rmd</code></a></small>
      <div class="d-none name"><code>full_documentation.Rmd</code></div>
    </div>

    
    
<style type="text/css"> 
  .centered-image{
    display: block; 
    margin-left: auto; 
    margin-right: auto;
    } 
    </style>
<div class="section level2">
<h2 id="plots-tab">Plots Tab<a class="anchor" aria-label="anchor" href="#plots-tab"></a>
</h2>
<p>The plots tab is used to explore and summarize the data using several
plot types. All plots may be created from either the full dataset or a
subset of cells based on metadata. For more, see the <a href="#Subsets">subsetting section</a>.</p>
<div class="section level3">
<h3 id="plot-switches">Plot Switches<a class="anchor" aria-label="anchor" href="#plot-switches"></a>
</h3>
<p>At the top of the options panel you will see several switches, one
for each plot type. To show/hide each plot, click the switch. The
available plot types are described briefly below, and each is described
in more detail in their respective sections.</p>
<ul>
<li><p><strong>DimPlot:</strong> shows each cell on a dimensional
reduction plot, e.g. a UMAP, tSNE, or other projection enabled by the
browser admin.</p></li>
<li><p><strong>Feature Plot:</strong> a DimPlot with cells colored by a
feature’s value (e.g. gene, surface protein, or other data enabled by
the browser admin).</p></li>
<li><p><strong>Violin Plot:</strong> displays the distribution of a
feature’s value across groups of cells as defined in the
metadata.</p></li>
<li><p><strong>Dot Plot:</strong> shows the average scaled value and
percentage of cells with non-zero values for multiple features across
groups of cells as defined by metadata categories.</p></li>
<li><p><strong>Scatterplot:</strong> compares values of two features.
<!--Correlation between values may be quantified in the [Gene Correlations Tab](#correlations-tab).--></p></li>
<li><p><strong>Ridge Plot:</strong> displays a density plot of a
feature’s value across a group of cells defined by metadata.</p></li>
<li><p><strong>Cell Proportion Plot:</strong> used to visualize the
proportion of cells belonging to each value of a categorical metadata
variable.</p></li>
<li><p><strong>Metadata Pie Chart:</strong> displays the number of
samples/patients in each selected categorical metadata value.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="feature-selection">Feature Selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h3>
<p>When plot types based on a gene or multiple genes are selected, the
feature entry menu will display beneath the switches. Any feature
(e.g. genes, surface proteins, gene signatures, or other data enabled by
the browser admin) may be entered here. As you type the name of a
feature, matching suggestions will appear along with the type of the
suggested feature.</p>
</div>
<div class="section level3">
<h3 id="Palettes">Palettes<a class="anchor" aria-label="anchor" href="#Palettes"></a>
</h3>
<p>Allows for the selection of color palettes for categorical and
continuous data. Selections made will apply to all plots in the plots
tab.</p>
</div>
<div class="section level3">
<h3 id="Subsets">Forming Subsets<a class="anchor" aria-label="anchor" href="#Subsets"></a>
</h3>
<p>Use the “Subset Options” menu to filter cells in the dataset based on
specified criteria, selection of which in these menus will include, not
exclude, cells defined by the criteria. These selections will be
inherited by all plots.</p>
<p>The dataset can be filtered by categorical metadata, feature
expression, or advanced subsetting (see <a href="#advanced-subsetting">Advanced (String) Subsetting</a>). Select
the “add filter” button to access the filter type menu. When choosing
categorical filter, any metadata variable can be chosen along with a
single value or combination of values. Select “select all” to include
all values in a filter, and select “deselect all” to clear all choices.
Once a filter has been chosen, the “confirm filter” button must be
selected before moving back to the filter menu. When choosing a feature
expression filter, the numeric filter menu offers the options to filter
cells in the dataset that express the chosen feature at levels above or
below a defined threshold, or within a specified range of expression.
Criteria defined by each menu are combined using <code>AND</code>
syntax, meaning that only cells with metadata selected in one menu, and
metadata selected in a second menu will be returned upon subsetting.
Menus for the other variables will update based on possible combinations
made in the current menu, and invalid combinations will appear in gray.
To select a value that is currently invalid, clear or edit selections
made in other filter menus, and select a different combination. Press
“Apply Subset” to create the subset from the current selection.</p>
<div class="section level4">
<h4 id="advanced-subsetting">Advanced (String) Subsetting<a class="anchor" aria-label="anchor" href="#advanced-subsetting"></a>
</h4>
<p>If the “Advanced Subsetting” option is chosen for filtering, a text
entry box will appear where subset criteria may be defined via R code.
This interface allows for the entry of additional types of criteria,
such as:</p>
<ul>
<li><p>Cells matching a threshold of numeric metadata
expression</p></li>
<li><p>Cells expressing a feature above or below a threshold</p></li>
<li><p>Cells with feature expression within a defined range</p></li>
</ul>
<p>For more information, see the <a href="advanced_subsetting_documentation.html">documentation for string
subsetting</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="DimPlots">Dimensional Reduction Plots<a class="anchor" aria-label="anchor" href="#DimPlots"></a>
</h3>
<p>Dimensional reduction plots are used to summarize relatedness between
cells across the dataset, e.g. using gene expression (for more, see <a href="scRNA_Plots_Explained.html">Interpereting scRNA-seq plots</a>).
Cells with similar profiles tend to cluster next to each other, but
quantitative connections can’t necessarily be made between distance on
the plot and similarity in gene expression (i.e. a cell that is 16 units
away from another cell is not twice as dissimilar from a cell that is 8
units away).</p>
<div class="section level4">
<h4 id="DimPlots-group-by">Metadata to Group By<a class="anchor" aria-label="anchor" href="#DimPlots-group-by"></a>
</h4>
<p>This menu is used to select a categorical metadata variable
(e.g. cell type) to be used to label cells by color. The key for colors
used will display in the legend to the right.</p>
</div>
<div class="section level4">
<h4 id="DimPlots-split-by">Metadata to Split By<a class="anchor" aria-label="anchor" href="#DimPlots-split-by"></a>
</h4>
<p>This menu sets a categorical metadata variable used to “split” cells
into separate plots for viewing side-by-side, and defaults to
“None”.</p>
<p>This menu is useful for comparing projections between different
metadata values. For example, broad differences in cell type proportions
may be visualized when splitting by disease response. These
obsservations can be analyzed in more detail in the differential gene
expression tab. <!-- and correlations tabs.--></p>
<p>When a variable is selected, a slider to choose the number of columns
will appear under the “Title options menu” (see <a href="#choose-number-of-columns">Choose number of columns</a> for more
info).</p>
</div>
<div class="section level4">
<h4 id="choose-projection">Choose Projection<a class="anchor" aria-label="anchor" href="#choose-projection"></a>
</h4>
<p>This menu is used to switch between different methods of dimensional
reduction: cells are plotted according to the coordinates selected in
the current projection method. Projection methods are generated during
dataset processing and added by the browser admin. Common projections
include UMAP and t-SNE. For more information on each projection included
with the current dataset, contact the user or organization that set up
the browser.</p>
</div>
<div class="section level4">
<h4 id="DimPlots-title-options">Title Options<a class="anchor" aria-label="anchor" href="#DimPlots-title-options"></a>
</h4>
<p>The title options menu is used to change or remove the title of the
dimensional reduction plot. When set to “Default” the title shown is the
name of the selected group by metadata variable. When set to “None”, no
title is shown. When set to “Custom”, an interface will appear beneath
the menu to enter a custom title. Press “Update” to apply changes to the
title, and “Reset” to revert the title to the default.</p>
<p><img src="full_documentation-custom_title_menu.png" class="centered-image" style="display: block; margin-left: auto; margin-right: auto;" width="336"></p>
</div>
<div class="section level4">
<h4 id="choose-number-of-columns">Choose Number of Columns<a class="anchor" aria-label="anchor" href="#choose-number-of-columns"></a>
</h4>
<p>This setting applies to split plots only, and remains hidden until a
selection is made in the “Choose Metadata to Split by” menu. Drag the
slider to specify the number of columns to use when displaying the
panels of a split plot. Possible values may be anywhere from 1 to the
number of panels created by the current split by variable.</p>
<p><img src="full_documentation-number_of_columns.png" class="centered-image" width="460"></p>
</div>
<div class="section level4">
<h4 id="label-groups">Label Groups<a class="anchor" aria-label="anchor" href="#label-groups"></a>
</h4>
<p>Toggles on and off labels displayed above each cluster of cells
according to the currently selected group by variable.</p>
</div>
<div class="section level4">
<h4 id="DimPlots-include-legend">Include Legend<a class="anchor" aria-label="anchor" href="#DimPlots-include-legend"></a>
</h4>
<p>Adds or removes legend displayed on the right side of the plot.</p>
<h4 id="DimPlots-include-legend-1">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#DimPlots-include-legend-1"></a>
</h4>
<p>When enabled, an interface appears allowing 
the user to change the height and width of the plot in pixels using either a slider bar or text entry. When 
entering a value in the text box, press return to update the plot with the 
value. Downloaded plots will have the dimensions defined here if "manually 
adjust plot dimensions" is checked. Otherwise defaults are...?</p>
<h4 id="DimPlots-include-legend-2">Download Button<a class="anchor" aria-label="anchor" href="#DimPlots-include-legend-2"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "DimPlot Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
</div>
<div class="section level3">
<h3 id="Feature-plots">Feature Plots<a class="anchor" aria-label="anchor" href="#Feature-plots"></a>
</h3>
<p>Feature plots are dimensional reduction plots that are colored
according to feature values instead of categorical metadata. Among other
things, feature plots can be used to see which cell types are expressing
a feature and to what extent, or to visualize differences in expression
between treatment groups when a split by metadata variable is
selected.</p>
<!-- ### Metadata to Group By

For feature plots, the group by selection menu controls the metadata to use for labeling cells, if ...?

This menu is used to select a categorical metadata variable to be used to coloring cells. Cells will be colored according to their associated metadata for the variable chosen, and the key for colors used will display in the legend to the right. -->
<div class="section level4">
<h4 id="Feature-plots-split-by">Metadata to Split By<a class="anchor" aria-label="anchor" href="#Feature-plots-split-by"></a>
</h4>
<p>This menu sets a categorical metadata variable used to “split” cells
into separate plots for viewing side-by-side. One panel will be created
for each value in the split by variable. If a single feature is entered
in the <a href="#feature-selection">feature selection</a> menu, the
number of columns used for the panels may be adjusted. If two or more
features are entered, however, the layout is fixed to place features by
row and split by groups by column.</p>
</div>
<div class="section level4">
<h4 id="choose-projection-1">Choose Projection<a class="anchor" aria-label="anchor" href="#choose-projection-1"></a>
</h4>
<p>This menu is used to switch between visualizing different methods of
dimensional reduction: cells are plotted according to the coordinates
selected in the current projection method. Projection methods are
calculated during dataset processing and added by the browser admin.
Common projections include UMAP and t-SNE. For more information on each
projection included with the current dataset, contact the user or
organization that set up the browser.</p>
</div>
<div class="section level4">
<h4 id="Feature-plots-title-options">Title Options<a class="anchor" aria-label="anchor" href="#Feature-plots-title-options"></a>
</h4>
<p>The title options menu is used to change or remove the titles
appearing above each panel of the feature plot. When set to “default”,
the titles shown are equal to the feature if one or multiple features
are entered without a split by variable, or to the values in the split
by metadata variable when one feature is entered with a split by
variable. When set to “none”, no titles are shown.</p>
<p>The “Custom” option is available in all settings except for when
multiple features are entered and a split by variable is chosen. When
set to “custom”, an interface will appear to customize titles for every
panel displayed on the screen. Enter values and press “Update” to apply
changes to the title. The
<i class="fas fa-rotate-right" role="presentation" aria-label="rotate-right icon"></i>
buttons to the right of each title entry will revert that individual
title to the default. Pressing “Reset” will revert the titles for all
panels to the default.</p>
</div>
<div class="section level4">
<h4 id="display-options-for-legend-title">Display Options for Legend Title<a class="anchor" aria-label="anchor" href="#display-options-for-legend-title"></a>
</h4>
<p>Used to control the title that displays above the legend. When set to
“Feature Name” (the default), the feature name will display above the
legend. When set to “Expression”, the legend will be set to
“Expression”. If “No Title” is selected, the title will be removed. This
menu is hidden when multiple features are entered and a split by
variable is chosen. In this case, no title will display above the
legend.</p>
</div>
<div class="section level4">
<h4 id="feature-choose-ncol">Choose Number of Columns<a class="anchor" aria-label="anchor" href="#feature-choose-ncol"></a>
</h4>
<p>This setting is available when one feature is selected with a split
by variable, or when multiple features are chosen without a split by
variable.</p>
<p>Drag the slider to specify the number of columns to distribute the
panels across. Possible values may be anywhere from 1 to the number of
panels. This setting is not available when feature co-expression is
enabled, or if multiple features are provided with a split by
variable.</p>
<p><img src="full_documentation-number_of_columns.png" class="centered-image" width="460"></p>
<p>If multiple features and a split by metadata variable are entered,
the output plot will be fixed in a layout of n splits by m features. If
you wish to have a custom layout for these plots, it is recommended to
create a separate plot for each feature with the desired layout,
download each plot, and then combine them using graphic design software.
A caveat of this approach is that the value scales may be different for
each plot.</p>
</div>
<div class="section level4">
<h4 id="share-scale-between-features">Share Scale Between Features<a class="anchor" aria-label="anchor" href="#share-scale-between-features"></a>
</h4>
<p>This setting is available when multiple features are entered with no
split by variable. When enabled, all features will be plotted using the
same scale. When disabled, each feature is plotted using a scale defined
by the minimum and maximum values for that feature. This setting must be
enabled to properly compare values between different features using a
color scale.</p>
</div>
<div class="section level4">
<h4 id="display-feature-name-above-panels">Display Feature Name Above Panels<a class="anchor" aria-label="anchor" href="#display-feature-name-above-panels"></a>
</h4>
<p>This setting is available when one feature is entered and split by
metadata is provided. When enabled, the feature name will display
centered above the panels created from the split by variable.</p>
</div>
<div class="section level4">
<h4 id="order-cells-by-expression">Order Cells by Expression<a class="anchor" aria-label="anchor" href="#order-cells-by-expression"></a>
</h4>
<p>When checked, cells with the highest values will be plotted on top of
the cells with lower expression values- essentially masking them. When
unchecked (the default), cells will be plotted in random order giving no
preference to either high or low value containing cells to be seen.
Enabling this setting can help with features that are expressed in only
a few cells in the dataset. Caution should be taken when visualizing
features with detectable values in many cells but low average values. In
these scenarios the values in the dataset may be exaggerated. To avoid
this, we recommend that feature plots be used alongside other means of
visualization, such as <a href="#violin-plots">violin plots</a>, <a href="#dot-plots">dot plots</a>, and <a href="#ridge-plots">ridge
plots</a>.</p>
</div>
<div class="section level4">
<h4 id="label-groups-1">Label Groups<a class="anchor" aria-label="anchor" href="#label-groups-1"></a>
</h4>
<p>Toggles group by variable labels above each set of cells on and
off.</p>
</div>
<div class="section level4">
<h4 id="feature-plots-include-legend">Include Legend<a class="anchor" aria-label="anchor" href="#feature-plots-include-legend"></a>
</h4>
<p>Toggles appearance of a legend to the right side of the plot on and
off.</p>
<h4 id="feature-plots-include-legend-1">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#feature-plots-include-legend-1"></a>
</h4>
<p>When enabled, an interface appears allowing 
the user to change the height and width of the plot in pixels using either a slider bar or text entry. When 
entering a value in the text box, press return to update the plot with the 
value. Downloaded plots will have the dimensions defined here if "manually 
adjust plot dimensions" is checked. Otherwise defaults are...?</p>
<h4 id="feature-plots-include-legend-2">Download Button<a class="anchor" aria-label="anchor" href="#feature-plots-include-legend-2"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "Feature Plot Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
</div>
<div class="section level3">
<h3 id="violin-plots">Violin Plots<a class="anchor" aria-label="anchor" href="#violin-plots"></a>
</h3>
<p><code># TODO: General description</code></p>
<div class="section level4">
<h4 id="violin-plots-group-by">Metadata for Primary Groups<a class="anchor" aria-label="anchor" href="#violin-plots-group-by"></a>
</h4>
<p>The categorical metadata variable chosen here will determine how
cells are grouped. One violin plot will be drawn for each unique value
in the chosen metadata variable.</p>
</div>
<div class="section level4">
<h4 id="violin-plots-split-by">Metadata for Secondary Grouping<a class="anchor" aria-label="anchor" href="#violin-plots-split-by"></a>
</h4>
<p>A second categorical variable may be supplied for more in-depth
comparison of groups. When specified, violins for each combination
between the values of the primary and secondary metadata variable will
be drawn. Violins will be colored according to the secondary variable,
and split according to the primary variable (see example below). In
general, the plot can be used to compare distributions for each value of
the primary variable between each value of the secondary variable.</p>
<p>[ADD IMAGE]</p>
</div>
<div class="section level4">
<h4 id="violin-plots-group-order">Order of Groups on Plot<a class="anchor" aria-label="anchor" href="#violin-plots-group-order"></a>
</h4>
<p>Sets the order groups specified by the primary metadata selection
appear on the plot. When set to “ascending”, values in the primary
metadata variable will display in alphanumeric order, from A to Z. When
set to “descending”, values will display in reverse alphanumeric order.
When set to “custom”, an interface will appear that will allow you to
arrange values in any order. To change the order of a value,
drag-and-drop the value in the interface.</p>
</div>
<div class="section level4">
<h4 id="violin-choose-ncol">Number of Columns<a class="anchor" aria-label="anchor" href="#violin-choose-ncol"></a>
</h4>
<p>This setting is available when multiple features have been
entered.</p>
<p>Drag the slider to specify the number of columns to distribute the
panels across. Possible values may be anywhere from 1 to the number of
panels. This setting is not available when feature co-expression is
enabled.</p>
<p><img src="full_documentation-number_of_columns.png" class="centered-image" width="460"></p>
</div>
<div class="section level4">
<h4 id="violin-plots-include-legend">Include Legend<a class="anchor" aria-label="anchor" href="#violin-plots-include-legend"></a>
</h4>
<p>When checked, the legend will display to the right-hand side of the
plot. when unchecked, the legend will be removed.</p>
<h4 id="violin-plots-include-legend-1">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#violin-plots-include-legend-1"></a>
</h4>
<p>When enabled, an interface appears allowing 
the user to change the height and width of the plot in pixels using either a slider bar or text entry. When 
entering a value in the text box, press return to update the plot with the 
value. Downloaded plots will have the dimensions defined here if "manually 
adjust plot dimensions" is checked. Otherwise defaults are...?</p>
<h4 id="violin-plots-include-legend-2">Download Button<a class="anchor" aria-label="anchor" href="#violin-plots-include-legend-2"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "Violin Plot Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
</div>
<div class="section level3">
<h3 id="dot-plots">Dot Plots<a class="anchor" aria-label="anchor" href="#dot-plots"></a>
</h3>
<pre><code><span><span class="co"># TODO: General description</span></span></code></pre>
<div class="section level4">
<h4 id="dot-plots-group-by">Metadata to group by<a class="anchor" aria-label="anchor" href="#dot-plots-group-by"></a>
</h4>
<p>Specifies the categorical metadata variable used to group cells. One
dot per feature will be created for each value of the selected
variable.</p>
</div>
<div class="section level4">
<h4 id="dot-plots-group-order">Order of Groups on Plot<a class="anchor" aria-label="anchor" href="#dot-plots-group-order"></a>
</h4>
<p>Sets the order groups appear on the plot. When set to “ascending”,
values in the group by metadata variable will display in ascending
alphanumeric order, from top to bottom. When set to “descending”, values
will display in descending alphanumeric order. When set to “custom”, an
interface will appear that will allow you to arrange values in any
order. To change the order of a value, drag-and-drop the value in the
interface.</p>
</div>
<div class="section level4">
<h4 id="dot-plots-include-legend">Include Legend<a class="anchor" aria-label="anchor" href="#dot-plots-include-legend"></a>
</h4>
<p>When checked, the legend will display to the right-hand side of the
plot. when unchecked, the legend will be removed.</p>
</div>
<div class="section level4">
<h4 id="manually-adjust-plot-dimensions">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#manually-adjust-plot-dimensions"></a>
</h4>
<p>When enabled, an interface appears allowing the user to change the
height and width of the plot. The height and width are entered in
pixels, and may be chosen using the slider or the text box. When
entering a value in the text box, press return to update the plot with
the value. Downloaded plots will have the dimensions defined here, if
“manually adjust plot dimensions” is checked.</p>
</div>
<div class="section level4">
<h4 id="use-separate-features-for-dot-plot">Use Separate Features for Dot Plot<a class="anchor" aria-label="anchor" href="#use-separate-features-for-dot-plot"></a>
</h4>
<p>When checked, an interface will appear to enter additional features,
or to remove features that were entered in the <a href="#feature-selection">feature selection</a> menu. Features added or
removed in this interface will apply only to the dot plot.</p>
<h4 id="use-separate-features-for-dot-plot-1">Download Button<a class="anchor" aria-label="anchor" href="#use-separate-features-for-dot-plot-1"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "Dot Plot Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
</div>
<div class="section level3">
<h3 id="scatterplots">Scatterplots<a class="anchor" aria-label="anchor" href="#scatterplots"></a>
</h3>
<p>Scatterplots are useful for viewing co-expression of two features.
All cells are plotted according to the expression of each feature, and
are useful for qualitative assessments of correlation.</p>
<div class="section level4">
<h4 id="feature-for-x-axis">Feature for x-axis<a class="anchor" aria-label="anchor" href="#feature-for-x-axis"></a>
</h4>
<p>A feature to plot on the x-axis. All features that can be entered in
the <a href="#feature-selection">feature selection</a> menu may be
entered here.</p>
</div>
<div class="section level4">
<h4 id="feature-for-y-axis">Feature for y-axis<a class="anchor" aria-label="anchor" href="#feature-for-y-axis"></a>
</h4>
<p>A feature to plot on the y-axis. All features that can be entered in
the <a href="#feature-selection">feature selection</a> menu may be
entered here.</p>
</div>
<div class="section level4">
<h4 id="scatterplots-group-by">Metadata to Group By<a class="anchor" aria-label="anchor" href="#scatterplots-group-by"></a>
</h4>
<p>This menu is used to select a categorical metadata variable to be
used to coloring cells. Cells will be colored according to their
associated metadata for the variable chosen, and the key for colors used
will display in the legend to the right.</p>
<!-- Include Legend -->
<h4 id="scatterplots-group-by-1">Include Legend<a class="anchor" aria-label="anchor" href="#scatterplots-group-by-1"></a>
</h4>
<p>Adds or removes legend displayed on the right side of the plot.</p>
</div>
<div class="section level4">
<h4 id="show-pearson-coefficient">Show Pearson Coefficient<a class="anchor" aria-label="anchor" href="#show-pearson-coefficient"></a>
</h4>
<p>When checked, a pearson correlation coefficient is computed between
the two features entered and displayed as the plot title. This is
checked by default.</p>
<h4 id="show-pearson-coefficient-1">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#show-pearson-coefficient-1"></a>
</h4>
<p>When enabled, an interface appears allowing 
the user to change the height and width of the plot in pixels using either a slider bar or text entry. When 
entering a value in the text box, press return to update the plot with the 
value. Downloaded plots will have the dimensions defined here if "manually 
adjust plot dimensions" is checked. Otherwise defaults are...?</p>
<h4 id="show-pearson-coefficient-2">Download Button<a class="anchor" aria-label="anchor" href="#show-pearson-coefficient-2"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "Scatterplot Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
</div>
<div class="section level3">
<h3 id="ridge-plots">Ridge Plots<a class="anchor" aria-label="anchor" href="#ridge-plots"></a>
</h3>
<div class="section level4">
<h4 id="ridge-plots-group-by">Metadata to Group by<a class="anchor" aria-label="anchor" href="#ridge-plots-group-by"></a>
</h4>
<p>Used to choose a categorical variable for splitting ridge plots into
groups. One ridge plot will be drawn for each value in the chosen
metadata variable. If set to “None”, a single ridge plot for all cells
will be plotted</p>
<!-- Include Legend -->
<h4 id="ridge-plots-group-by-1">Include Legend<a class="anchor" aria-label="anchor" href="#ridge-plots-group-by-1"></a>
</h4>
<p>Adds or removes legend displayed on the right side of the plot.</p>
</div>
<div class="section level4">
<h4 id="define-custom-x-axis-limits">Define Custom X-axis Limits<a class="anchor" aria-label="anchor" href="#define-custom-x-axis-limits"></a>
</h4>
<p>When checked, an interface will appear to change the X-axis limits.
To specify new limits, enter them in the text boxes for “Lower Bound”
and “Upper Bound” and press the
“<i class="fas fa-arrows-rotate" role="presentation" aria-label="arrows-rotate icon"></i>
Update” button to apply the new limits. To revert to the default limits,
press
“<i class="fas fa-arrow-rotate-right" role="presentation" aria-label="arrow-rotate-right icon"></i>
Reset”.</p>
<h4 id="define-custom-x-axis-limits-1">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#define-custom-x-axis-limits-1"></a>
</h4>
<p>When enabled, an interface appears allowing 
the user to change the height and width of the plot in pixels using either a slider bar or text entry. When 
entering a value in the text box, press return to update the plot with the 
value. Downloaded plots will have the dimensions defined here if "manually 
adjust plot dimensions" is checked. Otherwise defaults are...?</p>
<h4 id="define-custom-x-axis-limits-2">Download Button<a class="anchor" aria-label="anchor" href="#define-custom-x-axis-limits-2"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "Ridge Plot Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
<div class="section level4">
<h4 id="ridge-title-options">Title Options<a class="anchor" aria-label="anchor" href="#ridge-title-options"></a>
</h4>
<p>The title options menu is used to change or remove the title of the
plot. When set to “Default” the title shown is the name of the selected
feature. When set to “None”, no title is shown. When set to “Custom”, an
interface will appear beneath the menu to enter a custom title. Press
“Update” to apply changes to the title, and “Reset” to revert the title
to the default.</p>
</div>
</div>
<div class="section level3">
<h3 id="cell-proportion-plot">Cell Proportion Plot<a class="anchor" aria-label="anchor" href="#cell-proportion-plot"></a>
</h3>
<div class="section level4">
<h4 id="cell-proportion-plot-proportion-metadata">Choose Metadata for Proportions<a class="anchor" aria-label="anchor" href="#cell-proportion-plot-proportion-metadata"></a>
</h4>
<p>Used to choose the metadata to use for displaying cell type
proportions. The proportions illustrated by the stacked bar plots will
be relative to the number of cells matching each value in the chosen
metadata variable.</p>
</div>
<div class="section level4">
<h4 id="cell-proportion-plot-comparison-metadata">Choose Metadata for Proportion Comparison<a class="anchor" aria-label="anchor" href="#cell-proportion-plot-comparison-metadata"></a>
</h4>
<p>Used to choose a categorical metadata variable used for comparing
cell type proportions. One bar will be plotted for each value of the
chosen metadata variable.</p>
</div>
<div class="section level4">
<h4 id="cell-proportion-plot-title-options">Title Options<a class="anchor" aria-label="anchor" href="#cell-proportion-plot-title-options"></a>
</h4>
<p>The title options menu is used to change or remove the title of the
plot. When set to “Default” the title shown is the name of the selected
metadata variable for proportions. When set to “None”, no title is
shown. When set to “Custom”, an interface will appear beneath the menu
to enter a custom title. Press “Update” to apply changes to the title,
and “Reset” to revert the title to the default.</p>
<p><img src="full_documentation-custom_title_menu.png" class="centered-image" style="display: block; margin-left: auto; margin-right: auto;" width="336"></p>
</div>
<div class="section level4">
<h4 id="cell-proportion-plot-group-order">Order of Groups on Plot<a class="anchor" aria-label="anchor" href="#cell-proportion-plot-group-order"></a>
</h4>
<p>Sets the order bars created by the proportion comparison metadata
variable appear on the plot. When set to “ascending”, values in the
metadata variable will display in ascending alphanumeric order, from
left to right. When set to “descending”, values will display in
descending alphanumeric order. When set to “custom”, an interface will
appear that will allow you to arrange values in any order. To change the
order of a value, drag-and-drop the value in the interface.</p>
<!-- Manual Dimensions and Download Button -->
<h4 id="cell-proportion-plot-group-order-1">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#cell-proportion-plot-group-order-1"></a>
</h4>
<p>When enabled, an interface appears allowing 
the user to change the height and width of the plot in pixels using either a slider bar or text entry. When 
entering a value in the text box, press return to update the plot with the 
value. Downloaded plots will have the dimensions defined here if "manually 
adjust plot dimensions" is checked. Otherwise defaults are...?</p>
<h4 id="cell-proportion-plot-group-order-2">Download Button<a class="anchor" aria-label="anchor" href="#cell-proportion-plot-group-order-2"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "Cell Proportion Plot Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
</div>
<div class="section level3">
<h3 id="metadata-pie-chart">Metadata Pie Chart<a class="anchor" aria-label="anchor" href="#metadata-pie-chart"></a>
</h3>
<p>Metadata pie charts show metadata on the sample level rather than the
cell level, and is useful for summarizing the current dataset or subset
in this regard. Sample-level metadata is compiled based on a metadata
variable used as the sample (i.e., the metadata corresponding to “sample
ID”, “specimen”, “patient ID”, etc.), and is chosen by the user setting
up the browser.</p>
<div class="section level4">
<h4 id="view-number-of-patients-by-menu">View Number of Patients by Menu<a class="anchor" aria-label="anchor" href="#view-number-of-patients-by-menu"></a>
</h4>
<p>Sets a categorical metadata variable to use for displaying the sample
composition of the dataset. The size of each slice of the pie will be
determined by the number of samples matching the corresponding value of
a metadata variable provided to “View Number of Patients By”. If samples
match multiple values of the chosen metadata variable (for example, if
cell type is chosen and one sample has multiple cell types), the total
value of all slices will be greater than the total number of
patients.</p>
</div>
<div class="section level4">
<h4 id="title-options">Title Options<a class="anchor" aria-label="anchor" href="#title-options"></a>
</h4>
<p>The title options menu is used to change or remove the title of the
pie chart. When set to “Default” the title shown is the name of the
metadata variable in the “View Number of Patients by” menu. When set to
“None”, no title is shown. When set to “Custom”, an interface will
appear beneath the menu to enter a custom title. Press “Update” to apply
changes to the title, and “Reset” to revert the title to the
default.</p>
<p><img src="full_documentation-custom_title_menu.png" class="centered-image" style="display: block; margin-left: auto; margin-right: auto;" width="336"></p>
<!-- Include Legend -->
<h4 id="title-options-1">Include Legend<a class="anchor" aria-label="anchor" href="#title-options-1"></a>
</h4>
<p>Adds or removes legend displayed on the right side of the plot.</p>
<!-- Manual Dimensions -->
<h4 id="title-options-2">Manually Adjust Plot Dimensions<a class="anchor" aria-label="anchor" href="#title-options-2"></a>
</h4>
<p>When enabled, an interface appears allowing 
the user to change the height and width of the plot in pixels using either a slider bar or text entry. When 
entering a value in the text box, press return to update the plot with the 
value. Downloaded plots will have the dimensions defined here if "manually 
adjust plot dimensions" is checked. Otherwise defaults are...?</p>
<h4 id="title-options-3">Download Button<a class="anchor" aria-label="anchor" href="#title-options-3"></a>
</h4>
<p>Plots are downloaded by clicking the download button at the bottom left 
side of the "Metadata Pie Chart Specific Options" menu. A menu will appear 
with an option to choose whether to download a .png or .svg file. Downloads in the 
.svg format can be edited with Adobe Illustrator or a similar program.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="dge-tab">Differential Gene Expression Tab<a class="anchor" aria-label="anchor" href="#dge-tab"></a>
</h2>
<p>Differential gene expression (DGE) analysis compares groups of cells
based on their metadata, and returns a table of genes that differ in
expression between the groups. scExploreR uses <a href="https://www.biorxiv.org/content/10.1101/653253v1.full" class="external-link">presto</a>
for DGE analysis, which performs a Wilcoxon Rank Sum and an area under
the receiver operator curve (auROC) analysis on the expression data.</p>
<p>Two DGE testing modes are available: differential expression and
marker identification. These may be selected using the <strong>“Choose
test to perform”</strong> menu. When differential expression is
selected, two groups of cells are compared and the genes
upregulated/downregulated in one group vs. the other are shown. When
marker identification is selected, more than two groups will be
compared, and genes that serve as markers of each group (those that are
particularly upregulated or downregulated in that group) will be
displayed. When running differential gene expression, genes marked as
upregulated in a group are differentially expressed to the extent shown
in that group <strong>relative to the other group</strong>, whereas for
marker identification the genes upregulated in the marker class are
deferentially expressed to the extent shown in the marker class
<strong>relative to all other cells</strong>.</p>
<p>At this time, differential gene expression can only be performed for
genes. Other feature types will be supported in the future.</p>
<p>To run a DGE analysis, follow the process below. This process is
described in greater detail in subsequent sections, with more
information on the menus involved in each DGE mode.</p>
<ul>
<li><p>Determine if you would like to perform marker identification or
differential expression.</p></li>
<li><p>If performing marker identification, select the metadata to use
for marker classes. If performing differential expression, select two
groups based on metadata, or feature expression.</p></li>
<li><p>For either marker identification or differential expression,
determine if you would like to define groups using the full dataset, or
a subset of cells.</p></li>
</ul>
<div class="section level3">
<h3 id="marker-identification">Marker Identification Mode<a class="anchor" aria-label="anchor" href="#marker-identification"></a>
</h3>
<div class="section level4">
<h4 id="choose-marker-id-test">Choose Metadata to Use for Marker Identification<a class="anchor" aria-label="anchor" href="#choose-marker-id-test"></a>
</h4>
<p>The categorical metadata variable chosen here will be used as the
basis for determining groups for marker identification. The choices
available in “choose classes to include in marker computation” will
update based on the variable selected in this menu.</p>
</div>
<div class="section level4">
<h4 id="choose-classes-to-include-in-marker-computation">Choose classes to include in marker computation<a class="anchor" aria-label="anchor" href="#choose-classes-to-include-in-marker-computation"></a>
</h4>
<p>Specific categories from the metadata variable chosen in “choose
metadata to use for marker identification” may be included/excluded in
the analysis using this menu. The analysis will compute markers for each
of the categories selected in this menu. Groups (marker classes) are
defined in the context of the full dataset or a subset, depending on
selections made in the “<a href="#dge-subset-options">subset
options</a>” menu.</p>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression">Differential Expression Mode<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<p>The two groups defined in differential gene expression may be defined
in one of two ways: via categorical metadata, or via feature expression.
To define groups via feature expression, check <strong>“use feature
expression to define groups”</strong>. The groups are defined within the
context of the full dataset or a subset, depending on selections made in
the “<a href="#dge-subset-options">subset options</a>” menu.</p>
<div class="section level4">
<h4 id="groups-defined-by-categorical-metadata">Groups Defined by Categorical Metadata<a class="anchor" aria-label="anchor" href="#groups-defined-by-categorical-metadata"></a>
</h4>
<p>To define groups based on a categorical metadata variable, select a
variable in the <strong>“choose metadata to use for differential gene
expression”</strong>, and then choose a category to use for
<strong>“group 1”</strong> and <strong>“group 2”</strong>.</p>
</div>
<div class="section level4">
<h4 id="groups-defined-by-feature-expression">Groups Defined by Feature Expression<a class="anchor" aria-label="anchor" href="#groups-defined-by-feature-expression"></a>
</h4>
<p><img src="full_documentation-DGE_expr_threshold_widget.gif" style="float: right; margin-left: 15px;" width="195"></p>
<p>To define groups based on feature expression, enter a feature from
the search bar that appears after “use feature expression to define
groups” is checked.</p>
<p>When a feature is entered, an interface will appear with a plot of
the expression distribution of that feature in all cells of the dataset
(see right). Click the plot to choose an expression threshold. The cells
with expression values above the chosen threshold will be compared to
the cells below the threshold.</p>
<p>To aid in the selection of a threshold, an interface will display
below the plot with statistics on the chosen threshold. The interface
will display with the value of the chosen threshold, and the percentage
of cells above and below the threshold.</p>
</div>
</div>
<div class="section level3">
<h3 id="subset-options">Subset Options<a class="anchor" aria-label="anchor" href="#subset-options"></a>
</h3>
<p>The subset options menu is used to choose the context in which a
differential expression test is performed. If filter criteria are
entered in this interface, the test chosen in the menus above will be
performed only in cells that meet the chosen criteria.</p>
<p>For example, if a test based on feature expression is chosen, and a
specific sample is chosen in the subset options menus, then the
differential expression will compare cells above the expression
threshold chosen in that sample to cells below the expression threshold
in the sample.</p>
<p>The interface for entering criteria is the same as in the plots tab.
See <a href="#Subsets">“forming subsets”</a> for more info.</p>
</div>
<div class="section level3">
<h3 id="other-dge-settings">Other DGE Settings<a class="anchor" aria-label="anchor" href="#other-dge-settings"></a>
</h3>
<div class="section level4">
<h4 id="positive-markers-only">Positive Markers Only<a class="anchor" aria-label="anchor" href="#positive-markers-only"></a>
</h4>
<p>When checked, the table returned will only show genes that are
upregulated in each group/marker class. Disabling this option will also
show downregulated genes, but this information is redundant since genes
downregulated in a group will always show as being upregulated in
another group (i.e. in differential expression mode, a gene that is
expressed twice as much in group 1 compared to group 2 will be half as
expressed in group 2 relative to group 1).</p>
</div>
</div>
<div class="section level3">
<h3 id="interpreting-dge-results">Interpreting DGE Results<a class="anchor" aria-label="anchor" href="#interpreting-dge-results"></a>
</h3>
<p>After selecting a test and a subset, press update to run DGE on the
chosen groups. Depending on the size of the dataset and the performance
of the server, this may take up to two minutes to run, but usually
completes in under 30 seconds. When the test is complete, a report will
display in the main window.</p>
<p>The “test summary” section will display summary data on the test
selected and the subset it was performed on. The “test selected”
subsection will display either “differential expression” or “marker
identification” based on the test mode chosen. It will also display the
groups being compared if differential expression was chosen, or the
number of marker classes if marker identification was chosen. The
“subset used for test” subsection shows information on the subset of
cells chosen. For each categorical metadata variable shown in the app,
the categories represented by the cells in the test sample will be shown
(if cells from patients A, B, and C are present, this section will list
A, B, and C under “patients”). The number of cells in the subset tested
is also displayed, and the number of cells in each DGE group or marker
class is also shown.</p>
<div class="section level4">
<h4 id="differential-expression-columns">DGE Table<a class="anchor" aria-label="anchor" href="#differential-expression-columns"></a>
</h4>
<p>The columns of the DGE table are detailed below. The table may be
downloaded via the
“<i class="fas fa-table" role="presentation" aria-label="table icon"></i>
download table” button.</p>
<ul>
<li><p>Feature: the gene observed as being differentially
expressed.</p></li>
<li><p>Class (or group): the dge group, or marker class in which the
gene is differentially expressed.</p></li>
<li><p>Average expression: the average expression of the gene in the
group or class listed.</p></li>
<li><p>Log2FC: the log-2 fold change in expression of the gene in the
group or class, relative to either the other group (in differential
expression mode) or all other cells (in marker identification
mode).</p></li>
<li><p>AUC: the area under the receiver operator curve for the gene,
which serves as a measure of how well the gene serves as a marker for
the indicated group/class.</p></li>
<li><p>Adjusted p-value: the adjusted p-value associated with the log2FC
calculation in expression of the gene in the current group vs. the other
group, or all other cells.</p></li>
<li><p>Percent expression within class: the percentage of cells in the
indicated group or class that express the gene in the
group/class.</p></li>
<li><p>Percent expression outside class: the percentage of cells in the
indicated group or class that express the gene outside the
group/class.</p></li>
</ul>
<p>If either of the percent expression values are very low, or if there
is a large difference in percent expression inside the group
vs. outside, caution should be taken when interpreting the results.
Definitive comparisons are difficult to determine when fewer cells are
expressing the feature, and lower expression may be associated with the
<a href="https://doi.org/10.1038/nmeth.2967" class="external-link">dropout effect</a>.</p>
<ul>
<li>Additional info: a link to the <a href="https://www.genecards.org/" class="external-link">GeneCards</a> for each gene is
displayed here.</li>
</ul>
</div>
<div class="section level4">
<h4 id="dge-dimplot">DGE DimPlot<a class="anchor" aria-label="anchor" href="#dge-dimplot"></a>
</h4>
<p>To visually identify and verify the groups/marker classes being
compared in the selected test, a dimensional reduction plot will display
beneath the table. After the plot is displayed, a menu (“DimPlot
options”) will appear in the options panel on the left. The reduction
and the color by metadata can be changed here.</p>
<!-- The correlations tab was removed due to performance issues. The docs for the tab are left here but commented out.
# Correlations Tab {#correlations-tab}

This tab is used to compute the correlation between a gene entered and all other genes in the dataset. The correlations analysis uses a pearson correlation coefficient. Only genes may be used for computing correlation at this time: other feature types will be available in the future.

## Running a correlation analysis

To run an analysis, select a gene from the **"Gene Selection"** menu. All pearson coefficients returned will be for the genes relative to the selected gene. To view correlation results in a subset of cells rather than the full dataset, select categories for metadata variables in the **"Subset Options"** panel. The interface for entering subset criteria is the same as in the plots tab, and string subsetting may be performed for advanced subset selection. See ["forming subsets"](#Subsets) for more info. Press the **"submit"** button to run the analysis.

## Interpreting Correlation Results

The **"Subset Summary and Quality Statistics"** section will display the metadata included in the selected subset, along with quality statistics for the analysis. The number of cells in the selected subset will be displayed, along with the number and precentage of cells with non-zero reads for the currently selected gene. If either of these numbers are very low (less than 250 cells or so in the subset and less than 5% of cells with non-zero reads), caution should be taken when drawing conclusions from the results.

### Correlation table

The columns of the correlation table are as below. The table may be downloaded via the "`<i class="fas fa-table" role="presentation" aria-label="table icon"></i>`{=html} download table" button.

-   **Feature:** the gene being compared to the gene entered

-   **Correlation coefficient (global):** the pearson correlation coefficient of correlation with the gene entered, in all cells in the dataset

-   **Correlation coefficient (subset):** if a subset is entered in the options menu to the left, the pearson correlation coefficient of correlation in the chosen subset will display in this column. A column for both the full data and a subset will display.

### Correlation Scatterplots

Click any gene on the correlation table to view a scatterplot of expression for the gene entered vs. the selected gene (for more information on scatterplots, see the [scatterplots](#scatterplots) section). If a subset is selected, two plots will display, one for the full data and the other for the selected subset. The metadata variable used for coloring the plot may be selected using the **"choose metadata to group by"** menu. To download either of the plots, press the "`<i class="fas fa-square-poll-vertical" role="presentation" aria-label="square-poll-vertical icon"></i>`{=html} download scatterplot" button corresponding to the desired plot.
-->
</div>
</div>
</div>
<div class="section level2">
<h2 id="changing-datasets">Changing Datasets<a class="anchor" aria-label="anchor" href="#changing-datasets"></a>
</h2>
<p>To change the dataset, press the
<i class="far fa-elipsis-h" role="presentation" aria-label="elipsis-h icon"></i>
button in the upper right hand corner of the app window, and select
<strong>“Choose Dataset”</strong>. A window will appear with the
datasets that have been added to the browser, with a description of
each. Select a dataset and press “confirm selection” to load the
dataset. The dataset can be changed from any tab, and the loaded dataset
will affect all three tabs.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hematology Informatics Group.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
