<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Guide to Single-Cell Visualizations • scExploreR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Guide to Single-Cell Visualizations">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scExploreR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Help and Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/dataset_setup_walkthrough.html">App Setup Walkthrough</a></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">Exploring your Data</a></li>
    <li><a class="dropdown-item" href="../articles/object_guidelines.html">Object Preparation Guidelines</a></li>
    <li><a class="dropdown-item" href="../articles/docker.html">Deploying With Docker</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Additional Information</h6></li>
    <li><a class="dropdown-item" href="../articles/scRNA_Plots_Explained.html">Guide to scRNA-Seq Plots</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_subsetting_documentation.html">String Subsetting Guide</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Full Documentation</h6></li>
    <li><a class="dropdown-item" href="../articles/full_documentation.html">scExploreR Documentation</a></li>
    <li><a class="dropdown-item" href="../articles/config_documentation.html">Config App Documentation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/amc-heme/scExploreR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Guide to Single-Cell Visualizations</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amc-heme/scExploreR/blob/main/vignettes/scRNA_Plots_Explained.Rmd" class="external-link"><code>vignettes/scRNA_Plots_Explained.Rmd</code></a></small>
      <div class="d-none name"><code>scRNA_Plots_Explained.Rmd</code></div>
    </div>

    
    
<style type="text/css">
h1 {
  font-size: 1.17em;
}
</style>
<div class="section level2">
<h2 id="dimplot">Dimensional Reduction Plots (DimPlots)<a class="anchor" aria-label="anchor" href="#dimplot"></a>
</h2>
<p>Dimensional reduction plots (DimPlots) represent all cells according
to the whole of their transcriptional profile (or the profile of other
single-cell modalities such as surface protein expression or chromatin
accessibility). The whole of expression of one or more modalities is
computationally “reduced” to a two dimensional representation, and cells
with similar expression profiles appear close together on this
representation. The method for doing this varies and is chosen by the
bioinformatican that processed the currently loaded single-cell object,
though the standard method as of 2025 is principal component analysis
(PCA) followed by uniform manifold approximation and projection (UMAP).
All reduction methods implemented by the bioinformatican that prepared
the object can be toggled in the reductions menu, and you can contact
the person who prepared the object for more information (scExploreR
authors do not prepare the data shown in the app, only the app
itself).</p>
<p>Over interpretation of DimPlots is a common pitfall and should be
cautioned against. The structure of the data is dependent on numerous
variables including the specific algorithm used, data sparsity, and type
of data used to generate the plots (e.g. RNA-seq, ADT, etc). Therefore,
the distance between clusters cannot be used to draw biological
conclusions; all that can be concluded is that cells next to one another
have similar gene expression profiles.</p>
<p>For example, in the plot below, it’s not necessarily true that the
cells labeled as primitive cells are more different from CD4+ T Cells
than the cells labeled as NK cells, but it is reasonable to conclude
that CD4+ T Cells, CD8+ T Cells, and NK Cells have similar expression
profiles.</p>
<p><img src="scRNA_Plots_Explained_1_DimPlot_example.png" style="display: block; margin: 10px auto;" width="500"></p>
</div>
<div class="section level2">
<h2 id="feature">Feature Plots<a class="anchor" aria-label="anchor" href="#feature"></a>
</h2>
<p>Feature plots are DimPlots with cells colored according to the
expression of a feature, rather than according to categorical metadata.
A feature may be a gene, any other single-cell modality, or numeric
metadata. To view available feature types for the current single-cell
object, click the “What Can I Enter Here?” dropdown under the “Enter
features to display on plots” header.</p>
<p>Feature plots are useful for identifying clusters that express a gene
or surface protein to a greater extent, and when split by a metadata
variable, they can give an overview of how feature expression profiles
differ between different metadata groups.</p>
<p>Feature plots show a summary of expression, but they can over- or
under-represent the expression of features in the dataset. The order
cells are plotted may affect the representation of expression on the
plot, to a degree that varies based on the expression profile of the
feature. The “order cells” option toggles between plotting in random
order (the default), or plotting in order of expression, with the cells
with highest expression appearing above cells with lower expression.
Because of this, feature plots are best used as a summary, and should be
followed up with violin, ridge, and dot plots, or differential
expression analysis.</p>
</div>
<div class="section level2">
<h2 id="vln">Violin Plots<a class="anchor" aria-label="anchor" href="#vln"></a>
</h2>
<p>Violin plots show the distribution of expression values for feature
in groups of cells defined by a categorical metadata variable. The
distribution is represented as a “violin”, where the width of the violin
is determined by the number of cells in the group with expression values
matching each value on the y-axis. Cells are plotted over the violin as
points according to their expression values.</p>
</div>
<div class="section level2">
<h2 id="ridge">Ridge Plots<a class="anchor" aria-label="anchor" href="#ridge"></a>
</h2>
<p>Like violin plots, ridge plots represent the distribution of feature
expression. Ridge plots are density plots, where the height of the curve
is determined by the proportion of cells expressing a feature. Ridge
plots may represent expression in all cells, or in groups of cells
determined by a metadata variable.</p>
</div>
<div class="section level2">
<h2 id="dot">Dot Plots<a class="anchor" aria-label="anchor" href="#dot"></a>
</h2>
<p>Dot plots represent the average expression of features in cell groups
defined by a metadata variable, and the percentage of cells in each
group with non-zero expression of the feature. A dot is plotted for each
feature and metadata group combination: groups are plotted on rows, and
features are plotted on columns. The size of the dot shows the
percentage of cells in a metadata group expressing the feature it
represents, and the color of the dot shows the average expression of the
representative feature in the group. Dot plots are particularly
well-suited for visualizing large numbers of features at once.</p>
</div>
<div class="section level2">
<h2 id="scatter">Scatterplots<a class="anchor" aria-label="anchor" href="#scatter"></a>
</h2>
<p>Scatterplots visualize correlations in expression between two
features. To view a plot, enter two features in the “Scatterplot
Options” menu. Each point on the scatterplot represents a cell. Cells
are plotted with an x-coordinate equal to expression of the first
feature, and a y-coordinate equal to expression of the second feature.
The plot is similar in principle to a flow cytometry plot.</p>
<p>Scatterplots display a pearson correlation coefficient for expression
between the two features as the plot title. The pearson correlation
coefficient is based on the full dataset by default. To view correlation
in a cell type or a group of cells defined by a metadata variable,
subset the object for that group.</p>
<p>It is likely you will see many cells on both the x- and y-axis. This
indicates the dropout effect, a prevalent challenge in single-cell
sequencing. Due to the technical complexities of measuring mRNA and
other modalities in single-cells, it is common to find zero reads
matching a gene in a cell, even when the cell is in a population known
to express a gene. The dropout effect complicates statistical
conclusions that can be drawn from single-cell data.</p>
</div>
<div class="section level2">
<h2 id="proportion">Proportion Plots<a class="anchor" aria-label="anchor" href="#proportion"></a>
</h2>
<p>Proportion plots are used to compare cell type proportions across
groups of cells, typically according to sample level metadata. To view a
proportion plot, define the variable to use for proportions (typically
cell type), and the variable to use for comparing proportions (typically
sample or patient ID). Stacked bars will be drawn, one for each cell
group defined by the proportion comparison variable. The height of the
stacked bars within each group will be equal to the proportion of cells
within the comparison group matching each level of the proportion
variable.</p>
<p>Proportion plots are great for summarizing a dataset at the sample
level. For example, proportion plots can show the extent to which a
specific cell type is represented in different samples in the dataset,
and they can indicate samples with cell type proportions that differ
from most other samples in the dataset. Proportion plots can show a
continuum of samples from lowest to highest proportion of cells matching
a specific cell type, and the distribution of the proportion of that
cell type in the samples present.</p>
</div>
<div class="section level2">
<h2 id="pie">Metadata Pie Charts<a class="anchor" aria-label="anchor" href="#pie"></a>
</h2>
<p>Pie charts represent sample-level metadata, showing the number of
samples matching each group of a metadata variable. Pie charts only
display when the user preparing the dataset indicates a sample-level
variable to use for pie charts in the config app.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hematology Informatics Group.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
