<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analyzing Your Data in the Browser • scExploreR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing Your Data in the Browser">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scExploreR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Help and Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/dataset_setup_walkthrough.html">App Setup Walkthrough</a></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">Exploring your Data</a></li>
    <li><a class="dropdown-item" href="../articles/object_guidelines.html">Object Preparation Guidelines</a></li>
    <li><a class="dropdown-item" href="../articles/docker.html">Deploying With Docker</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Additional Information</h6></li>
    <li><a class="dropdown-item" href="../articles/scRNA_Plots_Explained.html">Guide to scRNA-Seq Plots</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_subsetting_documentation.html">String Subsetting Guide</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Full Documentation</h6></li>
    <li><a class="dropdown-item" href="../articles/full_documentation.html">scExploreR Documentation</a></li>
    <li><a class="dropdown-item" href="../articles/config_documentation.html">Config App Documentation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/amc-heme/scExploreR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Analyzing Your Data in the Browser</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amc-heme/scExploreR/blob/v0.9.6/vignettes/tutorial.Rmd" class="external-link"><code>vignettes/tutorial.Rmd</code></a></small>
      <div class="d-none name"><code>tutorial.Rmd</code></div>
    </div>

    
    
<style type="text/css"> 
  .centered-image{
    display: block; 
    margin-left: auto; 
    margin-right: auto;
    } 
    </style>
<div class="section level2">
<h2 id="scexplorer-walkthrough">scExploreR Walkthrough<a class="anchor" aria-label="anchor" href="#scexplorer-walkthrough"></a>
</h2>
<p>This tutorial will guide you through the use of scExploreR to analyze
your data. This is not a comprehensive exploration of each feature: for
more information on any feature, see the <a href="full_documentation.html">full documentation</a>. The browser used
for this tutorial is hosted here<!--[here]()-->. The dataset contains a
single bone marrow and peripheral blood sample from a healthy patient
(for more, see the <a href="https://figshare.com/articles/dataset/Expression_of_197_surface_markers_and_462_mRNAs_in_15281_cells_from_blood_and_bone_marrow_from_a_young_healthy_donor/13398065" class="external-link">download
page for the dataset</a>, and the related <a href="https://doi.org/10.1038/s41590-021-01059-0" class="external-link">publication</a>).</p>
<p>Key functionality is grouped into two <!--three--> tabs: plots and
differential expression. <!--and gene correlations.--> The interface on
each tab displays the plots and analysis results on the right, and the
options on the left hand panel (referred to in this document as the
options panel). The plots tab is selected by default, and is helpful for
summarizing data and producing visualizations. While this is a good
place to get an overview your dataset, some biological questions may be
best answered by starting in different tabs. This vignette will walk
through common biological questions and how to approach them using the
app.</p>
</div>
<div class="section level2">
<h2 id="in-which-cell-types-is-my-gene-deferentially-expressed">In which cell types is my gene deferentially expressed?<a class="anchor" aria-label="anchor" href="#in-which-cell-types-is-my-gene-deferentially-expressed"></a>
</h2>
<p>scExploreR provides differential gene analysis via a wilcoxon rank
sum test to determine which cell types express a gene to a greater
extent compared to the background (all cells in a dataset). To perform a
differential gene expression test, navigate to the “Differential
expression” tab.</p>
<p>Marker identification is used to determine which cell types your gene
is expressed in. This mode will divide all the cells in the dataset into
groups based on cell type, and will compare the expression of
<strong>each group</strong> to the <strong>average expression value for
all cells</strong>. This method is selected by default in the
<strong>“Choose Test to Perform”</strong> menu, so we will leave this
the same for now.</p>
<p>For <strong>Choose metadata to use for marker
identification</strong>, use the default setting, “Cell Type”. In other
datasets, this may be “clusters”, or something similar, and it is
usually the first option.</p>
<p>Under <strong>“Choose classes to include in marker
computation”</strong>, leave all cell types selected. If you are
interested in seeing data for a subset of only a few cell types, you may
select the types to include here, but <strong>note that the change in
expression will now be calculated relative to all cells within the cell
types entered, instead of all cells in the dataset</strong>. For
example, if only “CD8+ T Cells”, “B Cells”, and “NK Cells” are selected,
expression of each gene will be computed for each of these three cell
types, relative to all cells in the dataset that are either CD8+ T
Cells, B Cells, or NK cells.</p>
<p>Next, uncheck the <strong>“positive markers only”</strong> checkbox
at the bottom of the options panel. When this box is checked, only cell
types in which a gene is upregulated are included on the table, but when
unchecked, cell types where a gene is downregulated are also given.
Press <strong>“Update”</strong> to run the analysis.</p>
<p><img src="tutorial-001_Simple_DGE.gif" style="display: block; margin: 10px auto;" width="650"></p>
<div class="section level3">
<h3 id="interpreting-the-results">Interpreting the Results<a class="anchor" aria-label="anchor" href="#interpreting-the-results"></a>
</h3>
<p>After the analysis has finished computing, the results will display
in the main panel on the right. A summary of the test ran will display
on the top of the screen, with a description of the test performed, and
the metadata included in the subset (in this dataset, “Clusters (RNA
Based)” represents clusters calculated from the RNA assay, as opposed to
“Cell Type”, which is computed from both the RNA and surface protein
assays). Under the “number of cells per class” heading, be sure to
observe how many cells are in each cell type. The number will change if
tests are ran on a subset, and if there are too few cells represented,
differential expression results will be less reliable for that cell
type.</p>
<p><img src="tutorial-002_Results_Summary.png" class="centered-image" width="650"></p>
<p>Below the stats section, a table of differential gene expression
results is displayed. For details on each column of the table, see the
<a href="full_documentation.html#differential-expression-columns">Differential
Gene Expression</a> section of the full documentation. Enter a gene in
the text box below “Feature” to see results. For this example, we will
use “SPINK2”.</p>
<p>When determining if a gene is upregulated or downregulated in a
cluster, the “Log2FC”, “AUC”, and “Adjusted p-value” columns are the
most useful. The Log2FC will show if the gene is upregulated (positive
value) or downregulated (negative value), with the extent of the
expression change relative to all cells on a log-2 scale. Look at the
adjusted p-value column to see the Benjamini-Hochberg adjusted p-value
associated with this expression change, and the AUC value to view the
area under the receiving operator curve showing the performance of the
gene as a marker for the cell type indicated (how well the gene serves a
marker for a given cell type).</p>
<p><img src="tutorial-003_SPINK2_Exploration.gif" class="centered-image" width="650"></p>
<p>For SPINK2, the AUC value is greater than 0.7 in primitive cells,
meaning that the gene is a good marker for this cluster. This is
supported by the “percent expression within class” and “percent
expression outside class” columns, which show that the expression of
SPINK2 in the primitive cell types is much higher than in other cells
(48% of cells within the primitive cluster express this gene, whereas
2.8% of cells not in this cluster express it). This is also supported by
previous research identifying SPINK2 as a stem cell marker; this would
be likely to be upregulated in primitive cells, which are early in the
differentiation axis of PBMCs. For other cell types, the p-value appears
as significant, but the AUC values are much lower than for the primitive
cluster. The fold-change in expression is not very high, making SPINK2
an inconclusive marker for other cell types.</p>
</div>
</div>
<div class="section level2">
<h2 id="what-genes-are-markers-for-each-cell-type">What genes are markers for each cell type?<a class="anchor" aria-label="anchor" href="#what-genes-are-markers-for-each-cell-type"></a>
</h2>
<p>To identify markers, use the marker identification mode. The same
settings will be used as in the last example (cell type as the metadata
to use for marker identification, and all cell types selected). The
positive markers checkbox should again be left unchecked to see genes
downregulated in each cluster as well as those upregulated.</p>
<p><img src="tutorial-004_marker_id_settings.png" class="centered-image" width="203"></p>
<p>After clicking update to run the analysis, markers that differentiate
each cell type will be shown. To group markers by cell type, click the
“class” column header in the table. Clicking this header will toggle
sorting of the cell types by ascending or descending order. To view only
a particular cell type, type the name of that cell type in the text box
beneath the column header.</p>
<p>The list of markers can be filtered to view the most effective
markers for each cell type. To filter based on log-fold change, click
the text box beneath the “Log2FC” column. A slider will appear where the
upper and lower bounds of the values displayed can be adjusted. To enter
more precise bounds, you may type in this text box using the formula
<code>[lower bound] ... [upper bound]</code>. Either the lower bound or
the upper bound can be omitted to view markers that are above or below a
threshold, instead of using a range. For example, to view all markers
that are expressed at least 3 times higher on a log-2 fold scale, enter
<code>3 ...</code>. To view cells with a log-2 fold change of less than
zero, enter <code>... 0</code> (press return after entering the range).
You may also use the sliders that appear when clicking on the text box,
but a text entry allows for greater precision.</p>
<p><img src="tutorial-005_Filter_Demo_L2FC.gif" class="centered-image" width="650"></p>
<p>The same method can be used with the other columns. For example,
entering <code>0.85 ...</code> in the AUC column will subset for markers
that are particularly good at separating one cell class from another.
Sorting by class and subsetting via the text box will allow you to
quickly see the most effective markers for each cell type. To view more
information for each gene listed, click the GeneCards link in the
“additional info” column. When identifying markers, be sure to search
for each marker gene listed in the “Feature” text box to ensure that it
is not also a strong marker for other cell types. If it is, the gene
should be combined with other markers in the table specific to a given
cell type to properly distinguish it from other types.</p>
<p><img src="tutorial-006_Filter_Demo_AUC.gif" class="centered-image" width="650"></p>
<div class="section level3">
<h3 id="secondary-analysis-viewing-markers-in-plots-tab">Secondary Analysis: Viewing Markers in Plots tab<a class="anchor" aria-label="anchor" href="#secondary-analysis-viewing-markers-in-plots-tab"></a>
</h3>
<p>Markers can be visualized in the plots tab to support the analysis in
the DGE tab. After identifying a good marker for a cell class, copy the
marker name and go to the plots tab. In this example we will plot FCN1,
which is listed as a good marker for PBMC monocytes. Next enable a
feature plot, and type FCN1 into the <strong>“Enter features to display
on plots”</strong> text box.</p>
<p><img src="tutorial-007_secondary_plotting.gif" class="centered-image" width="650"></p>
<p>The resulting plot shows that FCN1 is highly expressed in the PBMC
monocytes, as well as in BM monocytes. This is to be expected since the
cell types are both monocytes, and were split due to them clustering
separately between the bone marrow and PBMC samples in this reference
object (for more information on the dataset, see <a href="https://figshare.com/articles/dataset/Expression_of_197_surface_markers_and_462_mRNAs_in_15281_cells_from_blood_and_bone_marrow_from_a_young_healthy_donor/13398065" class="external-link">here</a>).</p>
</div>
</div>
<div class="section level2">
<h2 id="marker-by-sample">How does a cell type vary between different samples?<a class="anchor" aria-label="anchor" href="#marker-by-sample"></a>
</h2>
<p>This reference dataset contains both blood and bone marrow samples
from a single patient (see <a href="https://figshare.com/articles/dataset/Expression_of_197_surface_markers_and_462_mRNAs_in_15281_cells_from_blood_and_bone_marrow_from_a_young_healthy_donor/13398065" class="external-link">here</a>
for more information). A UMAP of the dataset shows that dendritic cells
are present in both the blood and bone marrow samples, and that they
cluster separately. It would be useful to compare clusters between
datasets to see if there are genes uniquely expressed in each sample and
gain insight as to why this might be the case (determining if it is due
to unique biology or a technical artifact).</p>
<p><img src="tutorial-008_Dendritic_UMAP.png" class="centered-image" width="650"></p>
<p>To perform this comparison, choose “Marker Identification” under
<strong>“Choose Test to Perform”</strong>, and select “Sample Type”
under <strong>“Choose metadata to use for marker
identification”</strong> instead of “Cell Type”. The current selections
will compare all cells in the bone marrow sample to the blood sample. To
narrow this down to dendritic cells, go to <strong>“Subset
Options”</strong> and choose “Dendritic Cells” under <strong>“Subset by
Cell Type”</strong>. Leave the <strong>“Positive Markers Only”</strong>
checkbox unchecked to view negative markers as well as positive
ones.</p>
<p><img src="tutorial-009_Sample_Marker_ID_choices.gif" class="centered-image" width="650"></p>
<p>When viewing results, note in the <strong>“Test Selected”</strong>
and <strong>“Subset Used for Test”</strong> sections that the comparison
between the bone marrow (BM_200AB) and the PBMCs (PBMC_200AB) is for
dendritic cells only. To view strong positive markers, filter the AUC by
entering <code>0.85 ...</code> in the text box. To view strong negative
markers, filter by AUC using <code>... 0.15</code> (you may of course
adjust these bounds as desired).</p>
<p><img src="tutorial-009_Sample_Marker_ID_Pos_Neg_Markers.gif" class="centered-image" width="650"></p>
</div>
<div class="section level2">
<h2 id="how-do-i-compare-cells-based-on-genesurface-protein-expression">How do I compare cells based on gene/surface protein
expression?<a class="anchor" aria-label="anchor" href="#how-do-i-compare-cells-based-on-genesurface-protein-expression"></a>
</h2>
<p>The differential gene expression tab also allows you to group cells
by expression of a gene, by a surface protein, or by other feature types
present in your object. In this example, we will examine CD34, a marker
in hematopoiesis that tends to be expressed earlier in the
differentiation axis of PBMCs (see below; expression is greatest in the
“primitive” cluster). The top genes expressed in CD34+ cells vs. CD34-
cells will be examined.</p>
<p><img src="tutorial-010_CD34_Example.png" class="centered-image" width="650"></p>
<p>To perform this analysis, go to the differential gene expression tab
and select “Differential Expression” under the <strong>“Choose test to
perform”</strong> menu. Type “CD34”, and be sure to choose the entry
under the <strong>Surface Protein</strong> header (“CD34-AB (Surface
Protein)”). After selecting this option, a density plot of CD34
expression in the dataset will appear.</p>
<p><img src="tutorial-011_DGE_By_Expr_Demo.gif" class="centered-image" width="650"></p>
<p>On the plot, there are two peaks for CD34 expression. Most cells
appear either not to express CD34, or express it to an extent where it
is difficult to conclude if the expression reflects biological reality
or technical noise. A small portion of cells express CD34 to a greater
extent; these groups will be compared. To do this, click on the plot
between the two peaks. When mousing over the plot, a vertical line
giving the boundary used for comparison will be displayed: all cells
above this boundary will be compared to all cells below.</p>
<p><img src="tutorial-012_Selecting_Threshold.gif" class="centered-image" width="650"></p>
<p>After pressing <strong>“Update”</strong> to run the analysis, the
results will show genes expressed to a greater extent in CD34+ (high)
cells versus CD34- (low) ones. When sorting the table by AUC, several
good markers for both groups appear.</p>
<!--
# What genes in the dataset are correlated with my gene?

The correlations tab can be used, among other things, to determine which genes are co-expressed with a given gene in a sample. In this example, we will analyze the correlation of other genes with CXCR4, [a top marker for the peripheral blood dendritic cells](#marker-by-sample).

To run this analysis, switch to the correlation tab and enter "CXCR4" in the **"Gene Selection"** menu. Next, subset for dendritic cells by choosing this cell type under the **"Filter By Cell Type"** menus: this will allow for the comparison of correlation results in this cell type to the results in all cells. Press **"Submit"** to run the analysis.

![](tutorial-013_Correlation_Demo.gif){.centered-image width="650"}

The results will show the person correlation coefficients for both the full dataset, and the subset of dendritic cells. The analysis reveals several genes with a fairly strong positive correlation with CXCR4 expression in dendritic cells, but no genes have a strong positive or negative correlation in the context of the full dataset. Click a gene to view scatterplots of the correlation.

![](tutorial-014_Correlation_Scatterplots.gif){.centered-image width="650"}

To view correlation results for a particular gene of interest, you can search for it using the text box below the "Feature" column header on the table. -->
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hematology Informatics Group.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
